From f4ac96b17331570f1fbe7b90c696fe96180a0075 Mon Sep 17 00:00:00 2001
From: Matheus Sampaio Queiroga <srherobrine20@gmail.com>
Date: Fri, 28 Aug 2026 23:25:04 -0300
Subject: [PATCH] net: airoha: Refactor GDM device structure

Remove the intermediate `airoha_gdm_common` structure and integrate its
fields directly into `airoha_gdm_dev`. This simplifies the device
handling and aligns the data structures.

Signed-off-by: Matheus Sampaio Queiroga <srherobrine20@gmail.com>
---
 drivers/net/ethernet/airoha/airoha_eth.c | 1360 ++++++++++------------
 drivers/net/ethernet/airoha/airoha_eth.h |  115 +-
 drivers/net/ethernet/airoha/airoha_ppe.c |   14 +-
 3 files changed, 628 insertions(+), 861 deletions(-)

diff --git a/drivers/net/ethernet/airoha/airoha_eth.c b/drivers/net/ethernet/airoha/airoha_eth.c
index 0c8b7f156256..592d40e9984f 100644
--- a/drivers/net/ethernet/airoha/airoha_eth.c
+++ b/drivers/net/ethernet/airoha/airoha_eth.c
@@ -64,6 +64,15 @@ static void econet_prepare_qdma_cfg(struct airoha_qdma_mips_cfg *cfg,
 				    const struct airoha_eth_soc_data *soc,
 				    int id);
 
+static void airoha_mac_config(struct airoha_gdm_dev *dev, unsigned int mode,
+			      const struct phylink_link_state *state);
+static void airoha_mac_link_up(struct airoha_gdm_dev *dev,
+			       struct phy_device *phy,
+			       unsigned int mode, phy_interface_t interface,
+			       int speed, int duplex, bool tx_pause, bool rx_pause);
+static void airoha_mac_link_down(struct airoha_gdm_dev *dev, unsigned int mode,
+				 phy_interface_t interface);
+
 /* ------------------------------------------------------------------------- */
 /* Common Ethernet helpers shared by all supported frame engines. */
 
@@ -169,11 +178,11 @@ static void airoha_gdm_mac_config(struct phylink_config *config,
 				  unsigned int mode,
 				  const struct phylink_link_state *state)
 {
-	struct airoha_gdm_common *gdm;
+	struct airoha_gdm_dev *dev;
 
-	gdm = container_of(config, struct airoha_gdm_common, phylink_config);
-	if (gdm->mac_ops && gdm->mac_ops->mac_config)
-		gdm->mac_ops->mac_config(gdm->priv, mode, state);
+	dev = container_of(config, struct airoha_gdm_dev, phylink_config);
+	if (!airoha_is_econet(dev->eth))
+		airoha_mac_config(dev, mode, state);
 }
 
 static void airoha_gdm_mac_link_up(struct phylink_config *config,
@@ -183,23 +192,23 @@ static void airoha_gdm_mac_link_up(struct phylink_config *config,
 				   int speed, int duplex,
 				   bool tx_pause, bool rx_pause)
 {
-	struct airoha_gdm_common *gdm;
+	struct airoha_gdm_dev *dev;
 
-	gdm = container_of(config, struct airoha_gdm_common, phylink_config);
-	if (gdm->mac_ops && gdm->mac_ops->mac_link_up)
-		gdm->mac_ops->mac_link_up(gdm->priv, phy, mode, interface,
-					  speed, duplex, tx_pause, rx_pause);
+	dev = container_of(config, struct airoha_gdm_dev, phylink_config);
+	if (!airoha_is_econet(dev->eth))
+		airoha_mac_link_up(dev, phy, mode, interface, speed, duplex,
+				   tx_pause, rx_pause);
 }
 
 static void airoha_gdm_mac_link_down(struct phylink_config *config,
 				     unsigned int mode,
 				     phy_interface_t interface)
 {
-	struct airoha_gdm_common *gdm;
+	struct airoha_gdm_dev *dev;
 
-	gdm = container_of(config, struct airoha_gdm_common, phylink_config);
-	if (gdm->mac_ops && gdm->mac_ops->mac_link_down)
-		gdm->mac_ops->mac_link_down(gdm->priv, mode, interface);
+	dev = container_of(config, struct airoha_gdm_dev, phylink_config);
+	if (!airoha_is_econet(dev->eth))
+		airoha_mac_link_down(dev, mode, interface);
 }
 
 static const struct phylink_mac_ops airoha_gdm_phylink_ops = {
@@ -208,80 +217,61 @@ static const struct phylink_mac_ops airoha_gdm_phylink_ops = {
 	.mac_link_down = airoha_gdm_mac_link_down,
 };
 
-void airoha_gdm_common_init(struct airoha_gdm_common *gdm,
-			    struct airoha_eth *eth, struct net_device *netdev,
-			    enum airoha_eth_family family, u8 id,
-			    u8 pse_port, void *priv,
-			    const struct airoha_gdm_mac_ops *mac_ops)
-{
-	memset(gdm, 0, sizeof(*gdm));
-	gdm->magic = AIROHA_GDM_COMMON_MAGIC;
-	gdm->family = family;
-	gdm->id = id;
-	gdm->pse_port = pse_port;
-	gdm->eth = eth;
-	gdm->netdev = netdev;
-	gdm->priv = priv;
-	gdm->mac_ops = mac_ops;
-	gdm->phylink_config.dev = &netdev->dev;
-	gdm->phylink_config.type = PHYLINK_NETDEV;
-}
-
-int airoha_gdm_phylink_create(struct airoha_gdm_common *gdm,
+int airoha_gdm_phylink_create(struct airoha_gdm_dev *dev,
 			      struct device_node *np,
 			      phy_interface_t phy_mode)
 {
 	struct phylink *phylink;
 
-	phylink = phylink_create(&gdm->phylink_config,
+	phylink = phylink_create(&dev->phylink_config,
 				 of_fwnode_handle(np), phy_mode,
 				 &airoha_gdm_phylink_ops);
 	if (IS_ERR(phylink))
 		return PTR_ERR(phylink);
 
-	gdm->phylink = phylink;
+	dev->phylink = phylink;
 	return 0;
 }
 
-int airoha_gdm_phylink_connect(struct airoha_gdm_common *gdm,
+int airoha_gdm_phylink_connect(struct airoha_gdm_dev *dev,
 			       bool allow_no_phy)
 {
+	struct net_device *netdev = netdev_from_priv(dev);
 	int err;
 
-	if (!gdm->phylink)
+	if (!dev->phylink)
 		return -ENODEV;
 
-	err = phylink_of_phy_connect(gdm->phylink,
-				     gdm->netdev->dev.of_node, 0);
+	err = phylink_of_phy_connect(dev->phylink, netdev->dev.of_node, 0);
 	if (err) {
 		if (allow_no_phy && err == -ENODEV)
 			return 0;
 		return err;
 	}
 
-	phylink_start(gdm->phylink);
-	gdm->phylink_started = true;
+	phylink_start(dev->phylink);
+	dev->phylink_started = true;
 	return 0;
 }
 
-void airoha_gdm_phylink_disconnect(struct airoha_gdm_common *gdm)
+void airoha_gdm_phylink_disconnect(struct airoha_gdm_dev *dev)
 {
-	if (!gdm->phylink_started)
+	if (!dev->phylink_started)
 		return;
 
-	phylink_stop(gdm->phylink);
-	phylink_disconnect_phy(gdm->phylink);
-	gdm->phylink_started = false;
+	phylink_stop(dev->phylink);
+	phylink_disconnect_phy(dev->phylink);
+	dev->phylink_started = false;
 }
 
-void airoha_gdm_phylink_destroy(struct airoha_gdm_common *gdm)
+void airoha_gdm_phylink_destroy(struct airoha_gdm_dev *dev)
 {
-	airoha_gdm_phylink_disconnect(gdm);
-	if (!gdm->phylink)
+	airoha_gdm_phylink_disconnect(dev);
+	if (!dev->phylink)
 		return;
 
-	phylink_destroy(gdm->phylink);
-	gdm->phylink = NULL;
+	phylink_destroy(dev->phylink);
+	dev->phylink = NULL;
 }
 
 
@@ -1387,7 +1377,7 @@ static int econet_en751221_slm_init(struct airoha_qdma_mips *qdma,
 	return 0;
 }
 
-static int econet_init_hw_fwd(struct airoha_qdma_mips *qdma)
+static int econet_qdma_init_hfwd(struct airoha_qdma_mips *qdma)
 {
 	enum qregs_hwf_cfg_pkt_sz buf_size_cfg = QREGS_HWF_CFG_PKT_SZ_2048;
 	int ret, size, index, num_desc = qdma->cfg.num_fwd_descs;
@@ -1632,7 +1622,7 @@ static void econet_init_en751221_qdma(struct airoha_qdma_mips *qdma)
 	econet_wreg(cngst_cfg, &qdma->regs->tx_congest_cfg);
 }
 
-static int econet_init_final(struct airoha_qdma_mips *qdma)
+static int econet_qdma_hw_init(struct airoha_qdma_mips *qdma)
 {
 	struct qregs_qcfg qcfg;
 	int i;
@@ -1772,49 +1762,6 @@ static int econet_init_final(struct airoha_qdma_mips *qdma)
 	return 0;
 }
 
-static int econet_init(struct device *dev,
-		     struct airoha_qdma_mips *qdma,
-		     int *irqs)
-{
-	int err;
-	int i;
-
-	/* Make sure RX and TX are shutdown */
-	econet_wreg((struct qregs_qcfg) { 0 }, &qdma->regs->qdma_cfg);
-
-	err = econet_init_irqs(dev, qdma, irqs);
-	if (err)
-		return err;
-
-	for (i = 0; i < ARRAY_SIZE(qdma->q_rx); i++) {
-		err = econet_init_rx_queue(&qdma->q_rx[i], qdma,
-					 qdma->cfg.num_rx_descs[i]);
-		if (err)
-			return err;
-	}
-
-	for (i = 0; i < ARRAY_SIZE(qdma->q_tx_done); i++) {
-		err = econet_init_tx_doneq(&qdma->q_tx_done[i], qdma,
-					 qdma->cfg.done_list_size[i],
-					 qdma->cfg.done_list_irq_threshold[i]);
-		if (err)
-			return err;
-	}
-
-	for (i = 0; i < ARRAY_SIZE(qdma->q_tx); i++) {
-		err = econet_init_tx_queue(&qdma->q_tx[i], qdma,
-					 qdma->cfg.num_tx_descs[i]);
-		if (err)
-			return err;
-	}
-
-	err = econet_init_hw_fwd(qdma);
-	if (err)
-		return err;
-
-	return econet_init_final(qdma);
-}
-
 /* End init functions */
 
 static void econet_qdma_destroy_rxq_locked(struct econet_q_rx *q)
@@ -1966,6 +1913,55 @@ static int airoha_qdma_get_mips_irqs(struct platform_device *pdev,
 	return 0;
 }
 
+static int econet_qdma_init_irqs(struct platform_device *pdev,
+				 struct airoha_qdma_mips *qdma)
+{
+	int irqs[ECONET_MAX_QDMA_IRQS] = {};
+	int err;
+
+	err = airoha_qdma_get_mips_irqs(pdev, qdma->qdma, irqs);
+	if (err)
+		return err;
+
+	return econet_init_irqs(qdma->qdma->eth->dev, qdma, irqs);
+}
+
+static int econet_qdma_init_rx(struct airoha_qdma_mips *qdma)
+{
+	int i, err;
+
+	for (i = 0; i < ARRAY_SIZE(qdma->q_rx); i++) {
+		err = econet_init_rx_queue(&qdma->q_rx[i], qdma,
+					   qdma->cfg.num_rx_descs[i]);
+		if (err)
+			return err;
+	}
+
+	return 0;
+}
+
+static int econet_qdma_init_tx(struct airoha_qdma_mips *qdma)
+{
+	int i, err;
+
+	for (i = 0; i < ARRAY_SIZE(qdma->q_tx_done); i++) {
+		err = econet_init_tx_doneq(&qdma->q_tx_done[i], qdma,
+					   qdma->cfg.done_list_size[i],
+					   qdma->cfg.done_list_irq_threshold[i]);
+		if (err)
+			return err;
+	}
+
+	for (i = 0; i < ARRAY_SIZE(qdma->q_tx); i++) {
+		err = econet_init_tx_queue(&qdma->q_tx[i], qdma,
+					   qdma->cfg.num_tx_descs[i]);
+		if (err)
+			return err;
+	}
+
+	return 0;
+}
+
 /* Airoha QDMA backend. */
 static void airoha_qdma_set_irqmask(struct airoha_irq_bank *irq_bank,
 				    int index, u32 clear, u32 set)
@@ -2681,7 +2677,7 @@ static void airoha_qdma_cleanup_rx_queue(struct airoha_queue *q)
 			FIELD_PREP(RX_RING_DMA_IDX_MASK, q->tail));
 }
 
-static int airoha_qdma_init_rx(struct airoha_qdma *qdma)
+static int airoha_qdma_modern_init_rx(struct airoha_qdma *qdma)
 {
 	int i;
 
@@ -2966,7 +2962,7 @@ static int airoha_qdma_tx_irq_init(struct airoha_tx_irq_queue *irq_q,
 	return 0;
 }
 
-static int airoha_qdma_init_tx(struct airoha_qdma *qdma)
+static int airoha_qdma_modern_init_tx(struct airoha_qdma *qdma)
 {
 	int i, err;
 
@@ -3036,7 +3032,7 @@ void airoha_qdma_cleanup_tx_queue(struct airoha_queue *q)
 	spin_unlock_bh(&q->lock);
 }
 
-static int airoha_qdma_init_hfwd_queues(struct airoha_qdma *qdma)
+static int airoha_qdma_modern_init_hfwd(struct airoha_qdma *qdma)
 {
 	int size, index, num_desc = HW_DSCP_NUM;
 	struct airoha_eth *eth = qdma->eth;
@@ -3212,7 +3208,7 @@ static void airoha_qdma_init_qos_stats(struct airoha_qdma *qdma)
 	}
 }
 
-static int airoha_qdma_hw_init(struct airoha_qdma *qdma)
+static int airoha_qdma_modern_hw_init(struct airoha_qdma *qdma)
 {
 	int i;
 
@@ -3345,8 +3341,8 @@ static irqreturn_t airoha_irq_handler(int irq, void *dev_instance)
 	return IRQ_HANDLED;
 }
 
-static int airoha_qdma_init_irq_banks(struct platform_device *pdev,
-				      struct airoha_qdma *qdma)
+static int airoha_qdma_modern_init_irqs(struct platform_device *pdev,
+					struct airoha_qdma *qdma)
 {
 	struct airoha_eth *eth = qdma->eth;
 	int i, id = qdma - &eth->qdma[0];
@@ -3382,67 +3378,108 @@ static int airoha_qdma_init_irq_banks(struct platform_device *pdev,
 	return 0;
 }
 
-int airoha_qdma_init(struct platform_device *pdev,
-		     struct airoha_eth *eth, struct airoha_qdma *qdma)
+static int airoha_qdma_prepare(struct platform_device *pdev,
+			       struct airoha_eth *eth,
+			       struct airoha_qdma *qdma, int id)
 {
-	int err, id = qdma - &eth->qdma[0];
-	const char *res;
-
-	res = devm_kasprintf(eth->dev, GFP_KERNEL, "qdma%d", id);
-	if (!res)
-		return -ENOMEM;
-
-	qdma->regs = devm_platform_ioremap_resource_byname(pdev, res);
-	if (IS_ERR(qdma->regs))
-		return dev_err_probe(eth->dev, PTR_ERR(qdma->regs),
-				     "failed to iomap qdma%d regs\n", id);
-
-	if (airoha_is_econet(eth)) {
+	if (airoha_qdma_is_legacy(eth)) {
 		struct airoha_qdma_mips_cfg cfg;
 		struct airoha_qdma_mips *econet;
-		int irqs[ECONET_MAX_QDMA_IRQS] = {};
 
 		econet_prepare_qdma_cfg(&cfg, eth->soc, id);
-		airoha_qdma_setup(qdma, eth, qdma->regs, id, cfg.num_channels);
-
-		err = airoha_qdma_get_mips_irqs(pdev, qdma, irqs);
-		if (err)
-			return err;
-
 		econet = devm_kzalloc(eth->dev, sizeof(*econet), GFP_KERNEL);
 		if (!econet)
 			return -ENOMEM;
 
+		airoha_qdma_setup(qdma, eth, qdma->regs, id, cfg.num_channels);
 		qdma->econet = econet;
 		econet->qdma = qdma;
 		econet->regs = qdma->regs;
+		econet->cfg = cfg;
 		mutex_init(&econet->lock);
-		memcpy(&econet->cfg, &cfg, sizeof(cfg));
 
-		err = econet_init(eth->dev, econet, irqs);
-		if (err) {
-			econet_qdma_destroy(econet);
-			qdma->econet = NULL;
-		}
+		/* Keep RX/TX stopped while the common init sequence sets up QDMA. */
+		econet_wreg((struct qregs_qcfg) { 0 }, &econet->regs->qdma_cfg);
 
-		return err;
+		return 0;
 	}
 
-	airoha_qdma_setup(qdma, eth, qdma->regs, id, AIROHA_NUM_QOS_CHANNELS);
-
-	qdma->q_tx = devm_kcalloc(&pdev->dev,
-				 eth->soc->tx_ring,
-				 sizeof(*qdma->q_tx), GFP_KERNEL);
+	qdma->q_tx = devm_kcalloc(&pdev->dev, eth->soc->tx_ring,
+				  sizeof(*qdma->q_tx), GFP_KERNEL);
 	if (!qdma->q_tx)
 		return -ENOMEM;
 
-	qdma->q_rx = devm_kcalloc(&pdev->dev,
-				 eth->soc->rx_ring,
-				 sizeof(*qdma->q_rx), GFP_KERNEL);
+	qdma->q_rx = devm_kcalloc(&pdev->dev, eth->soc->rx_ring,
+				  sizeof(*qdma->q_rx), GFP_KERNEL);
 	if (!qdma->q_rx)
 		return -ENOMEM;
 
-	err = airoha_qdma_init_irq_banks(pdev, qdma);
+	airoha_qdma_setup(qdma, eth, qdma->regs, id, AIROHA_NUM_QOS_CHANNELS);
+
+	return 0;
+}
+
+static int airoha_qdma_init_irqs(struct platform_device *pdev,
+				 struct airoha_qdma *qdma)
+{
+	if (qdma->econet)
+		return econet_qdma_init_irqs(pdev, qdma->econet);
+
+	return airoha_qdma_modern_init_irqs(pdev, qdma);
+}
+
+static int airoha_qdma_init_rx(struct airoha_qdma *qdma)
+{
+	if (qdma->econet)
+		return econet_qdma_init_rx(qdma->econet);
+
+	return airoha_qdma_modern_init_rx(qdma);
+}
+
+static int airoha_qdma_init_tx(struct airoha_qdma *qdma)
+{
+	if (qdma->econet)
+		return econet_qdma_init_tx(qdma->econet);
+
+	return airoha_qdma_modern_init_tx(qdma);
+}
+
+static int airoha_qdma_init_hfwd(struct airoha_qdma *qdma)
+{
+	if (qdma->econet)
+		return econet_qdma_init_hfwd(qdma->econet);
+
+	return airoha_qdma_modern_init_hfwd(qdma);
+}
+
+static int airoha_qdma_hw_init(struct airoha_qdma *qdma)
+{
+	if (qdma->econet)
+		return econet_qdma_hw_init(qdma->econet);
+
+	return airoha_qdma_modern_hw_init(qdma);
+}
+
+int airoha_qdma_init(struct platform_device *pdev,
+		     struct airoha_eth *eth, struct airoha_qdma *qdma)
+{
+	int err, id = qdma - &eth->qdma[0];
+	const char *res;
+
+	res = devm_kasprintf(eth->dev, GFP_KERNEL, "qdma%d", id);
+	if (!res)
+		return -ENOMEM;
+
+	qdma->regs = devm_platform_ioremap_resource_byname(pdev, res);
+	if (IS_ERR(qdma->regs))
+		return dev_err_probe(eth->dev, PTR_ERR(qdma->regs),
+				     "failed to iomap qdma%d regs\n", id);
+
+	err = airoha_qdma_prepare(pdev, eth, qdma, id);
+	if (err)
+		return err;
+
+	err = airoha_qdma_init_irqs(pdev, qdma);
 	if (err)
 		return err;
 
@@ -3454,7 +3491,7 @@ int airoha_qdma_init(struct platform_device *pdev,
 	if (err)
 		return err;
 
-	err = airoha_qdma_init_hfwd_queues(qdma);
+	err = airoha_qdma_init_hfwd(qdma);
 	if (err)
 		return err;
 
@@ -3615,49 +3652,6 @@ void airoha_qdma_stop(struct airoha_qdma *qdma)
 	}
 }
 
-/* ------------------------------------------------------------------------- */
-/* EcoNet GDM/netdev datapath. */
-
-/* OMCI rules may be referenced by more than one UNI/VEIP cookie. */
-
-static int econet_xpon_start(struct airoha_gdm_dev *port);
-static void econet_xpon_stop(struct airoha_gdm_dev *port);
-
-static void econet_set_macaddr(struct airoha_gdm_dev *port, const u8 *addr)
-{
-	struct gdm_mymac_msb msb;
-	struct gdm_mymac_lsb lsb;
-
-	scoped_guard(spinlock, &port->reg_lock) {
-		msb = econet_rreg(&port->econet_regs->mymac_msb);
-		lsb = econet_rreg(&port->econet_regs->mymac_lsb);
-		set_gdm_mymac_msb_a(&msb, addr[0]);
-		set_gdm_mymac_msb_b(&msb, addr[1]);
-		set_gdm_mymac_lsb_c(&lsb, addr[2]);
-		set_gdm_mymac_lsb_d(&lsb, addr[3]);
-		set_gdm_mymac_lsb_e(&lsb, addr[4]);
-		set_gdm_mymac_lsb_f(&lsb, addr[5]);
-		econet_wreg(lsb, &port->econet_regs->mymac_lsb);
-		econet_wreg(msb, &port->econet_regs->mymac_msb);
-	}
-
-}
-
-static int econet_dev_set_macaddr(struct net_device *dev, void *p)
-{
-	struct airoha_gdm_dev *port = netdev_priv(dev);
-	int err;
-
-	err = eth_mac_addr(dev, p);
-	if (err)
-		return err;
-
-	econet_set_macaddr(port, dev->dev_addr);
-
-	return 0;
-}
-
-
 static u16 econet_gdm_oversize_len(struct airoha_gdm_dev *port, int mtu)
 {
 	u16 len = ETH_HLEN + mtu + ETH_FCS_LEN;
@@ -3700,142 +3694,6 @@ static void econet_set_gdm_port_fwd_cfg(struct airoha_gdm_dev *port,
 	econet_wreg(fc, &port->econet_regs->fwd_cfg);
 }
 
-static int econet_dev_init(struct net_device *dev)
-{
-	struct airoha_gdm_dev *port = netdev_priv(dev);
-
-	econet_set_macaddr(port, dev->dev_addr);
-	/* Route each GDM port to the CPU side of its matching QDMA. */
-	econet_set_gdm_port_fwd_cfg(port,
-		port->fport == ETX_FPORT_GDM2 ?
-		ETX_FPORT_QDMA1_CPU : ETX_FPORT_QDMA0_CPU);
-
-	if (airoha_is_econet(port->eth) &&
-	    port->fport == ETX_FPORT_GDM1) {
-		struct g1_cport_cfg cport_cfg;
-		struct gdm_vlan vlan;
-
-		/*
-		 * Match macSetMACCR() in the EN7512 vendor driver. CDMA1
-		 * needs the 0x8100 insertion TPID and the GDM1 CPORT padding
-		 * path is always enabled before LAN traffic is started. Do not
-		 * rely on the bootloader leaving either register initialized.
-		 */
-		scoped_guard(spinlock, &port->reg_lock) {
-			vlan = econet_rreg(&port->econet_regs->vlan);
-			vlan.tpid = ETH_P_8021Q;
-			econet_wreg(vlan, &port->econet_regs->vlan);
-
-			cport_cfg = econet_rreg(&port->econet_regs->g1_cport_cfg);
-			set_gdm_g1_cport_cfg_pad(&cport_cfg, true);
-			econet_wreg(cport_cfg, &port->econet_regs->g1_cport_cfg);
-		}
-	}
-
-	return 0;
-}
-
-static int econet_dev_open(struct net_device *dev)
-{
-	struct airoha_gdm_dev *port = netdev_priv(dev);
-	bool dsa = netdev_uses_dsa(dev);
-	struct gdm_len_th rlt;
-	int err;
-
-	err = airoha_gdm_phylink_connect(&port->common,
-					 (port->flags & AIROHA_PRIV_F_XPON_MANAGED));
-	if (err) {
-		netdev_err(dev, "could not attach PHY: %d\n", err);
-		return err;
-	}
-	if ((port->flags & AIROHA_PRIV_F_XPON_MANAGED))
-		netif_carrier_off(dev);
-
-	/* The MT7530 CPU port is represented by ethernet = <&gdm1>. Keep
-	 * the MTK special tag on the DSA conduit and disable it on a direct
-	 * PHY/WAN GDM. EN751221 additionally needs the legacy CDM/GDM bits
-	 * used by the vendor special-tag datapath.
-	 */
-	scoped_guard(spinlock, &port->reg_lock) {
-		if (airoha_is_econet(port->eth) &&
-		    port->fport == ETX_FPORT_GDM1) {
-			struct fwd_cfg fc = econet_rreg(&port->econet_regs->fwd_cfg);
-
-			if (dsa) {
-				fc.word &= ~EN751221_GDM_UNTAG_EN;
-				fc.word |= EN751221_GDM_STAG_EN;
-			} else {
-				fc.word &= ~EN751221_GDM_STAG_EN;
-			}
-			econet_wreg(fc, &port->econet_regs->fwd_cfg);
-
-			/* CDMA_CSG_CFG is at offset 0 from the GDM1 window. */
-			airoha_rmw(port->regs, 0, EN751221_CDM_STAG_EN,
-				   dsa ? EN751221_CDM_STAG_EN : 0);
-		}
-
-		econet_wreg((u32)dsa, &port->econet_regs->stag_en);
-		rlt = econet_rreg(&port->econet_regs->rx_len_threshold);
-		set_gdm_len_th_runt_len(&rlt, 60);
-		set_gdm_len_th_oversize_len(&rlt,
-					 econet_gdm_oversize_len(port, dev->mtu));
-		econet_wreg(rlt, &port->econet_regs->rx_len_threshold);
-	}
-
-	airoha_qdma_start(port->qdma);
-	netif_tx_start_all_queues(dev);
-
-	err = econet_xpon_start(port);
-	if (err) {
-		netif_tx_disable(dev);
-		airoha_qdma_stop(port->qdma);
-		airoha_gdm_phylink_disconnect(&port->common);
-		return err;
-	}
-
-	return 0;
-}
-
-static int econet_dev_stop(struct net_device *dev)
-{
-	struct airoha_gdm_dev *port = netdev_priv(dev);
-
-	netif_tx_disable(dev);
-	econet_xpon_stop(port);
-	airoha_gdm_phylink_disconnect(&port->common);
-
-	scoped_guard(spinlock, &port->reg_lock) {
-		econet_wreg(0U, &port->econet_regs->stag_en);
-		if (airoha_is_econet(port->eth) &&
-		    port->fport == ETX_FPORT_GDM1) {
-			struct fwd_cfg fc = econet_rreg(&port->econet_regs->fwd_cfg);
-
-			fc.word &= ~EN751221_GDM_STAG_EN;
-			econet_wreg(fc, &port->econet_regs->fwd_cfg);
-			airoha_rmw(port->regs, 0, EN751221_CDM_STAG_EN, 0);
-		}
-	}
-
-	airoha_qdma_stop(port->qdma);
-	return 0;
-}
-
-static int econet_dev_change_mtu(struct net_device *dev, int mtu)
-{
-	struct airoha_gdm_dev *port = netdev_priv(dev);
-	struct gdm_len_th rlt;
-
-	guard(spinlock)(&port->reg_lock);
-	rlt = econet_rreg(&port->econet_regs->rx_len_threshold);
-	set_gdm_len_th_oversize_len(&rlt, econet_gdm_oversize_len(port, mtu));
-	econet_wreg(rlt, &port->econet_regs->rx_len_threshold);
-
-	WRITE_ONCE(dev->mtu, mtu);
-
-	return 0;
-}
-
-
 static int econet_validate_xpon_gdm2(struct net_device *netdev,
 				     struct airoha_gdm_dev **gdm)
 {
@@ -3933,7 +3791,7 @@ static int econet_xpon_classify(struct airoha_gdm_dev *port,
 					  true, pcp, info);
 }
 
-static netdev_tx_t econet_dev_xmit(struct sk_buff *skb, struct net_device *dev)
+static netdev_tx_t econet_qdma_xmit(struct sk_buff *skb, struct net_device *dev)
 {
 	struct airoha_gdm_dev *port = netdev_priv(dev);
 	struct airoha_qdma_skb_meta skb_meta;
@@ -4107,207 +3965,34 @@ static void econet_update_hw_stats(struct airoha_gdm_dev *port)
 		port->stats.rx_drops += econet_rreg(&c->rx.g2.edrops);
 		port->stats.rx_broadcast += econet_rreg(&c->rx.g2.bcast);
 		port->stats.rx_multicast += econet_rreg(&c->rx.g2.mcast);
-		port->stats.rx_crc_error += econet_rreg(&c->rx.g2.ecrc);
-		port->stats.rx_fragment += econet_rreg(&c->rx.g2.efrag);
-		port->stats.rx_jabber += econet_rreg(&c->rx.g2.ejabber);
-
-		i = 0;
-		port->stats.rx_len[i] += econet_rreg(&c->rx.g2.f_less_64);
-		port->stats.rx_len[i++] += econet_rreg(&c->rx.g2.f_64);
-
-		port->stats.rx_len[i++] += econet_rreg(&c->rx.g2.f_65_127);
-		port->stats.rx_len[i++] += econet_rreg(&c->rx.g2.f_128_255);
-		port->stats.rx_len[i++] += econet_rreg(&c->rx.g2.f_256_511);
-		port->stats.rx_len[i++] += econet_rreg(&c->rx.g2.f_512_1023);
-		port->stats.rx_len[i++] += econet_rreg(&c->rx.g2.f_1024_1518);
-		port->stats.rx_len[i++] += econet_rreg(&c->rx.g2.f_more_1518);
-	} else {
-		port->stats.rx_drops += econet_rreg(&c->rx.drops_err);
-		port->stats.rx_drops += econet_rreg(&c->rx.drops_fc);
-		port->stats.rx_drops += econet_rreg(&c->rx.drops_rc);
-		port->stats.rx_drops += econet_rreg(&c->rx.drops_overflow);
-	}
-
-	set_gdm_cl_cnt_rx(&cc, true);
-	set_gdm_cl_cnt_tx(&cc, true);
-	econet_wreg(cc, &port->econet_regs->clear_counters);
-
-	u64_stats_update_end(&port->stats.syncp);
-}
-
-static void econet_dev_get_stats64(struct net_device *dev,
-				 struct rtnl_link_stats64 *storage)
-{
-	struct airoha_gdm_dev *port = netdev_priv(dev);
-	unsigned int start;
-
-	econet_update_hw_stats(port);
-	do {
-		start = u64_stats_fetch_begin(&port->stats.syncp);
-		storage->rx_packets = port->stats.rx_ok_pkts;
-		storage->tx_packets = port->stats.tx_ok_pkts;
-		storage->rx_bytes = port->stats.rx_ok_bytes;
-		storage->tx_bytes = port->stats.tx_ok_bytes;
-		storage->rx_errors = port->stats.rx_errors;
-		storage->rx_dropped = port->stats.rx_drops;
-		storage->tx_dropped = port->stats.tx_drops;
-		storage->rx_over_errors = port->stats.rx_over_errors;
-		if (port->g2_stats) {
-			storage->multicast = port->stats.rx_multicast;
-			storage->rx_crc_errors = port->stats.rx_crc_error;
-		}
-	} while (u64_stats_fetch_retry(&port->stats.syncp, start));
-}
-
-static void econet_ethtool_get_mac_stats(struct net_device *dev,
-				       struct ethtool_eth_mac_stats *stats)
-{
-	struct airoha_gdm_dev *port = netdev_priv(dev);
-	unsigned int start;
-
-	econet_update_hw_stats(port);
-
-	if (!port->g2_stats)
-		return;
-
-	do {
-		start = u64_stats_fetch_begin(&port->stats.syncp);
-		stats->MulticastFramesXmittedOK = port->stats.tx_multicast;
-		stats->BroadcastFramesXmittedOK = port->stats.tx_broadcast;
-		stats->BroadcastFramesReceivedOK = port->stats.rx_broadcast;
-	} while (u64_stats_fetch_retry(&port->stats.syncp, start));
-}
-
-static const struct ethtool_rmon_hist_range econet_ethtool_rmon_ranges[] = {
-	{    0,    64 },
-	{   65,   127 },
-	{  128,   255 },
-	{  256,   511 },
-	{  512,  1023 },
-	{ 1024,  1518 },
-	{ 1519, 10239 },
-	{},
-};
-
-static void
-econet_ethtool_get_rmon_stats(struct net_device *dev,
-			    struct ethtool_rmon_stats *stats,
-			    const struct ethtool_rmon_hist_range **ranges)
-{
-	struct airoha_gdm_dev *port = netdev_priv(dev);
-	struct airoha_hw_stats *hw_stats = &port->stats;
-	unsigned int start;
-
-	BUILD_BUG_ON(ARRAY_SIZE(econet_ethtool_rmon_ranges) !=
-		     ARRAY_SIZE(hw_stats->tx_len) + 1);
-	BUILD_BUG_ON(ARRAY_SIZE(econet_ethtool_rmon_ranges) !=
-		     ARRAY_SIZE(hw_stats->rx_len) + 1);
-
-	*ranges = econet_ethtool_rmon_ranges;
-	econet_update_hw_stats(port);
-
-	if (!port->g2_stats)
-		return;
-
-	do {
-		int i;
-
-		start = u64_stats_fetch_begin(&port->stats.syncp);
-		stats->fragments = hw_stats->rx_fragment;
-		stats->jabbers = hw_stats->rx_jabber;
-		for (i = 0; i < ARRAY_SIZE(econet_ethtool_rmon_ranges) - 1;
-		     i++) {
-			stats->hist[i] = hw_stats->rx_len[i];
-			stats->hist_tx[i] = hw_stats->tx_len[i];
-		}
-	} while (u64_stats_fetch_retry(&port->stats.syncp, start));
-}
-
-static int econet_dev_setup_tc(struct net_device *dev,
-			       enum tc_setup_type type, void *type_data)
-{
-	struct airoha_gdm_dev *port = netdev_priv(dev);
-
-	return airoha_ppe_dev_setup_tc(port->common.ppe, dev, type,
-				       type_data);
-}
-
-static const struct net_device_ops econet_netdev_ops = {
-	.ndo_init		= econet_dev_init,
-	.ndo_open		= econet_dev_open,
-	.ndo_stop		= econet_dev_stop,
-	.ndo_change_mtu		= econet_dev_change_mtu,
-	.ndo_start_xmit		= econet_dev_xmit,
-	.ndo_get_stats64        = econet_dev_get_stats64,
-	.ndo_set_mac_address	= econet_dev_set_macaddr,
-	.ndo_setup_tc		= econet_dev_setup_tc,
-};
-
-static int econet_ethtool_get_link_ksettings(struct net_device *dev,
-					     struct ethtool_link_ksettings *cmd)
-{
-	struct airoha_gdm_dev *port = netdev_priv(dev);
-	struct airoha_xpon_link_state state;
-	unsigned long flags;
-
-	if (!(READ_ONCE(port->flags) & AIROHA_PRIV_F_XPON_MANAGED))
-		return phylink_ethtool_ksettings_get(port->common.phylink, cmd);
-
-	spin_lock_irqsave(&port->xpon_state_lock, flags);
-	state = port->xpon_link;
-	spin_unlock_irqrestore(&port->xpon_state_lock, flags);
-
-	if (!state.valid && port->common.phylink_started)
-		return phylink_ethtool_ksettings_get(port->common.phylink, cmd);
-
-	ethtool_link_ksettings_zero_link_mode(cmd, supported);
-	ethtool_link_ksettings_zero_link_mode(cmd, advertising);
-	ethtool_link_ksettings_zero_link_mode(cmd, lp_advertising);
-	linkmode_set_bit(ETHTOOL_LINK_MODE_FIBRE_BIT,
-			 cmd->link_modes.supported);
-	linkmode_set_bit(ETHTOOL_LINK_MODE_FIBRE_BIT,
-			 cmd->link_modes.advertising);
-
-	cmd->base.speed = state.valid ? state.speed : SPEED_UNKNOWN;
-	cmd->base.duplex = state.valid ? state.duplex : DUPLEX_UNKNOWN;
-	cmd->base.autoneg = AUTONEG_DISABLE;
-	cmd->base.port = state.valid ? state.port : PORT_FIBRE;
-	cmd->base.phy_address = 0xff;
-
-	return 0;
-}
-
-static int econet_ethtool_set_link_ksettings(struct net_device *dev,
-					     const struct ethtool_link_ksettings *cmd)
-{
-	struct airoha_gdm_dev *port = netdev_priv(dev);
-
-	if ((READ_ONCE(port->flags) & AIROHA_PRIV_F_XPON_MANAGED))
-		return -EOPNOTSUPP;
+		port->stats.rx_crc_error += econet_rreg(&c->rx.g2.ecrc);
+		port->stats.rx_fragment += econet_rreg(&c->rx.g2.efrag);
+		port->stats.rx_jabber += econet_rreg(&c->rx.g2.ejabber);
 
-	return phylink_ethtool_ksettings_set(port->common.phylink, cmd);
-}
+		i = 0;
+		port->stats.rx_len[i] += econet_rreg(&c->rx.g2.f_less_64);
+		port->stats.rx_len[i++] += econet_rreg(&c->rx.g2.f_64);
 
-static int econet_ethtool_nway_reset(struct net_device *dev)
-{
-	struct airoha_gdm_dev *port = netdev_priv(dev);
+		port->stats.rx_len[i++] += econet_rreg(&c->rx.g2.f_65_127);
+		port->stats.rx_len[i++] += econet_rreg(&c->rx.g2.f_128_255);
+		port->stats.rx_len[i++] += econet_rreg(&c->rx.g2.f_256_511);
+		port->stats.rx_len[i++] += econet_rreg(&c->rx.g2.f_512_1023);
+		port->stats.rx_len[i++] += econet_rreg(&c->rx.g2.f_1024_1518);
+		port->stats.rx_len[i++] += econet_rreg(&c->rx.g2.f_more_1518);
+	} else {
+		port->stats.rx_drops += econet_rreg(&c->rx.drops_err);
+		port->stats.rx_drops += econet_rreg(&c->rx.drops_fc);
+		port->stats.rx_drops += econet_rreg(&c->rx.drops_rc);
+		port->stats.rx_drops += econet_rreg(&c->rx.drops_overflow);
+	}
 
-	if ((READ_ONCE(port->flags) & AIROHA_PRIV_F_XPON_MANAGED))
-		return -EOPNOTSUPP;
+	set_gdm_cl_cnt_rx(&cc, true);
+	set_gdm_cl_cnt_tx(&cc, true);
+	econet_wreg(cc, &port->econet_regs->clear_counters);
 
-	return phylink_ethtool_nway_reset(port->common.phylink);
+	u64_stats_update_end(&port->stats.syncp);
 }
 
-static const struct ethtool_ops econet_ethtool_ops = {
-	.get_drvinfo		= airoha_eth_get_drvinfo,
-	.get_link		= ethtool_op_get_link,
-	.get_link_ksettings	= econet_ethtool_get_link_ksettings,
-	.set_link_ksettings	= econet_ethtool_set_link_ksettings,
-	.nway_reset		= econet_ethtool_nway_reset,
-	.get_eth_mac_stats      = econet_ethtool_get_mac_stats,
-	.get_rmon_stats		= econet_ethtool_get_rmon_stats,
-};
-
 /* Frame-engine platform state is shared by all supported SoCs. */
 
 static struct airoha_gdm_dev *
@@ -4339,7 +4024,7 @@ int econet_rx_before_recv(struct airoha_eth *eth, struct sk_buff *skb,
 	if (IS_ERR(dev))
 		return PTR_ERR(dev);
 
-	port = dev->common.netdev;
+	port = netdev_from_priv(dev);
 
 	/*
 	 * EN7512/EN7521 keeps the MT7530 special tag in-band. The vendor
@@ -4354,29 +4039,29 @@ int econet_rx_before_recv(struct airoha_eth *eth, struct sk_buff *skb,
 	return 0;
 }
 
-static int econet_xpon_start(struct airoha_gdm_dev *port)
+static int airoha_gdm_xpon_start(struct airoha_gdm_dev *dev)
 {
 	int ret = 0;
 
-	mutex_lock(&port->xpon_lock);
-	if (port->xpon_ops && !port->xpon_started) {
-		ret = port->xpon_ops->start(port->xpon_priv);
+	mutex_lock(&dev->xpon_lock);
+	if (dev->xpon_ops && !dev->xpon_started) {
+		ret = dev->xpon_ops->start(dev->xpon_priv);
 		if (!ret)
-			port->xpon_started = true;
+			dev->xpon_started = true;
 	}
-	mutex_unlock(&port->xpon_lock);
+	mutex_unlock(&dev->xpon_lock);
 
 	return ret;
 }
 
-static void econet_xpon_stop(struct airoha_gdm_dev *port)
+static void airoha_gdm_xpon_stop(struct airoha_gdm_dev *dev)
 {
-	mutex_lock(&port->xpon_lock);
-	if (port->xpon_ops && port->xpon_started) {
-		port->xpon_started = false;
-		port->xpon_ops->stop(port->xpon_priv);
+	mutex_lock(&dev->xpon_lock);
+	if (dev->xpon_ops && dev->xpon_started) {
+		dev->xpon_started = false;
+		dev->xpon_ops->stop(dev->xpon_priv);
 	}
-	mutex_unlock(&port->xpon_lock);
+	mutex_unlock(&dev->xpon_lock);
 }
 
 static int econet_set_xpon_mode(struct net_device *netdev,
@@ -4516,7 +4201,7 @@ static int econet_register_xpon(struct net_device *netdev,
 		return ret;
 
 	if (netif_running(netdev))
-		return econet_xpon_start(port);
+		return airoha_gdm_xpon_start(port);
 
 	return 0;
 }
@@ -4537,7 +4222,7 @@ static void econet_unregister_xpon(struct net_device *netdev,
 	 * before the provider private pointer can disappear.
 	 */
 	airoha_qdma_mips_set_xpon_irq(port->qdma, port->xpon_mode, false);
-	econet_xpon_stop(port);
+	airoha_gdm_xpon_stop(port);
 
 	mutex_lock(&port->xpon_lock);
 	if (port->xpon_ops == ops && port->xpon_priv == priv) {
@@ -4929,7 +4614,7 @@ bool econet_rx_xpon_oam(struct airoha_eth *eth, u8 qdma_id,
 	if (raw0 & ERX_XPON_CRC_ERROR)
 		flags |= AIROHA_XPON_OAM_RX_F_CRC_ERROR;
 
-	skb->dev = port->common.netdev;
+	skb->dev = netdev_from_priv(port);
 	skb_len = skb->len;
 	atomic64_inc(&port->xpon_oam_rx_packets);
 	atomic64_add(skb_len, &port->xpon_oam_rx_bytes);
@@ -5064,16 +4749,6 @@ static void econet_prepare_qdma_cfg(struct airoha_qdma_mips_cfg *cfg,
 	}
 }
 
-
-/* ------------------------------------------------------------------------- */
-/* Airoha GDM/QDMA datapath. */
-
-static const struct net_device_ops airoha_netdev_ops;
-
-static int airoha_qdma_set_gpon_dba_report(struct net_device *netdev,
-					   int channel, bool enable);
-
-
 static int airoha_set_macaddr(struct airoha_gdm_dev *dev, const u8 *addr)
 {
 	u8 ref_addr[ETH_ALEN] __aligned(2);
@@ -5081,6 +4756,26 @@ static int airoha_set_macaddr(struct airoha_gdm_dev *dev, const u8 *addr)
 	u32 reg, val, lmin, lmax;
 	int i;
 
+	if (eth->soc->mac_addr_mode == AIROHA_MAC_ADDR_GDM_MASK) {
+		struct gdm_mymac_msb msb = {};
+		struct gdm_mymac_lsb lsb = {};
+
+		set_gdm_mymac_msb_lsb_mask(&msb, 0xf8);
+		set_gdm_mymac_msb_a(&msb, addr[0]);
+		set_gdm_mymac_msb_b(&msb, addr[1]);
+		set_gdm_mymac_lsb_c(&lsb, addr[2]);
+		set_gdm_mymac_lsb_d(&lsb, addr[3]);
+		set_gdm_mymac_lsb_e(&lsb, addr[4]);
+		set_gdm_mymac_lsb_f(&lsb, addr[5]);
+
+		scoped_guard(spinlock, &dev->reg_lock) {
+			econet_wreg(lsb, &dev->econet_regs->mymac_lsb);
+			econet_wreg(msb, &dev->econet_regs->mymac_msb);
+		}
+
+		return 0;
+	}
+
 	eth_zero_addr(ref_addr);
 	lmin = (addr[3] << 16) | (addr[4] << 8) | addr[5];
 	lmax = lmin;
@@ -5150,16 +4845,15 @@ static int airoha_validate_xpon_gdm2(struct net_device *netdev,
 					 struct airoha_gdm_dev **gdm)
 {
 	struct airoha_gdm_dev *dev;
-	struct airoha_eth *eth;
 
 	if (!netdev)
 		return -EINVAL;
-	if (netdev->netdev_ops != &airoha_netdev_ops)
+
+	dev = airoha_gdm_from_netdev(netdev);
+	if (!dev || !dev->eth || airoha_is_econet(dev->eth))
 		return -ENODEV;
 
-	dev = netdev_priv(netdev);
-	eth = dev->eth;
-	if (!eth || !dev->port || !airoha_is(eth, airoha_en7523) ||
+	if (!dev->port || !airoha_is(dev->eth, airoha_en7523) ||
 	    dev->port->id != AIROHA_GDM2_IDX)
 		return -EOPNOTSUPP;
 
@@ -5294,6 +4988,9 @@ static int airoha_xpon_set_datapath(struct net_device *netdev,
 	return 0;
 }
 
+static int airoha_qdma_set_gpon_dba_report(struct net_device *netdev,
+					   int channel, bool enable);
+
 static int airoha_xpon_set_tcont_channel(struct net_device *netdev,
 				      unsigned int channel, bool enable)
 {
@@ -5350,30 +5047,6 @@ static int airoha_xpon_set_tcont_channel(struct net_device *netdev,
 	return 0;
 }
 
-static int airoha_eth_xpon_start(struct airoha_gdm_dev *dev)
-{
-	int ret = 0;
-
-	mutex_lock(&dev->xpon_lock);
-	if (dev->xpon_ops && !dev->xpon_started) {
-		ret = dev->xpon_ops->start(dev->xpon_priv);
-		if (!ret)
-			dev->xpon_started = true;
-	}
-	mutex_unlock(&dev->xpon_lock);
-
-	return ret;
-}
-
-static void airoha_eth_xpon_stop(struct airoha_gdm_dev *dev)
-{
-	mutex_lock(&dev->xpon_lock);
-	if (dev->xpon_ops && dev->xpon_started) {
-		dev->xpon_started = false;
-		dev->xpon_ops->stop(dev->xpon_priv);
-	}
-	mutex_unlock(&dev->xpon_lock);
-}
 
 static int airoha_xpon_register_link(struct net_device *netdev,
 			     enum airoha_xpon_mode mode,
@@ -5412,7 +5085,7 @@ static int airoha_xpon_register_link(struct net_device *netdev,
 		return ret;
 
 	if (netif_running(netdev))
-		return airoha_eth_xpon_start(dev);
+		return airoha_gdm_xpon_start(dev);
 
 	return 0;
 }
@@ -5427,7 +5100,7 @@ static void airoha_xpon_unregister_link(struct net_device *netdev,
 	if (airoha_validate_xpon_gdm2(netdev, &dev))
 		return;
 
-	airoha_eth_xpon_stop(dev);
+	airoha_gdm_xpon_stop(dev);
 	mutex_lock(&dev->xpon_lock);
 	if (dev->xpon_ops == ops && dev->xpon_priv == priv) {
 		dev->xpon_ops = NULL;
@@ -6370,21 +6043,6 @@ static int airoha_hw_init(struct platform_device *pdev,
 	if (err)
 		goto error;
 
-	/* GDM netdevs on the Airoha path are allocated before PPE so an
-	 * nvmem -EPROBE_DEFER cannot leave DMA running. Bind the shared PPE
-	 * frontend once hardware initialization has completed.
-	 */
-	for (i = 0; i < eth->soc->max_gdm_ports; i++) {
-		struct airoha_gdm_port *port = eth->ports ? eth->ports[i] : NULL;
-		int j;
-
-		if (!port)
-			continue;
-		for (j = 0; j < ARRAY_SIZE(port->devs); j++)
-			if (port->devs[j])
-				port->devs[j]->common.ppe = eth->ppe_dev;
-	}
-
 	set_bit(DEV_STATE_INITIALIZED, &eth->state);
 	return 0;
 
@@ -6573,6 +6231,19 @@ static void airoha_update_hw_stats(struct airoha_gdm_dev *dev)
 	spin_unlock(&port->lock);
 }
 
+static void airoha_gdm_update_hw_stats(struct airoha_gdm_dev *dev)
+{
+	if (airoha_is_econet(dev->eth))
+		econet_update_hw_stats(dev);
+	else
+		airoha_update_hw_stats(dev);
+}
+
+static bool airoha_gdm_has_extended_stats(struct airoha_gdm_dev *dev)
+{
+	return !airoha_is_econet(dev->eth) || dev->g2_stats;
+}
+
 static void airoha_update_netdev_features(struct airoha_gdm_dev *dev)
 {
 	struct airoha_eth *eth = dev->eth;
@@ -6696,36 +6367,37 @@ static int airoha_dev_open(struct net_device *netdev)
 	struct airoha_gdm_dev *dev = netdev_priv(netdev);
 	struct airoha_gdm_port *port = dev->port;
 	struct airoha_eth *eth = dev->eth;
+	bool econet = airoha_is_econet(dev->eth);
 	struct airoha_qdma *qdma;
 	u32 pse_port = FE_PSE_PORT_PPE1;
-	int err, qdma_id;
+	int err;
 
-	/* HW LRO is configured on the QDMA and it is shared between all the
-	 * devices using it. Opening a second device on the same QDMA turns HW
-	 * GRO off rather than refusing to open: on an xPON board GDM2 and the
-	 * WAN share QDMA1, and failing here leaves the PON interface down.
-	 */
 	qdma = airoha_qdma_deref(dev);
-	qdma_id = qdma - &eth->qdma[0];
+	if (!econet) {
+		int qdma_id = qdma - &eth->qdma[0];
 
-	if (qdma->users && airoha_fe_lro_is_enabled(eth, qdma_id)) {
-		netdev_info(netdev, "disabling HW GRO on shared QDMA%d\n",
-			    qdma_id);
-		airoha_fe_lro_disable(eth, qdma_id);
+		/* HW LRO is configured on the QDMA and it is shared between all
+		 * the devices using it. Opening a second device on the same QDMA
+		 * turns HW GRO off rather than refusing to open.
+		 */
+		if (qdma->users && airoha_fe_lro_is_enabled(eth, qdma_id)) {
+			netdev_info(netdev,
+				    "disabling HW GRO on shared QDMA%d\n", qdma_id);
+			airoha_fe_lro_disable(eth, qdma_id);
+		}
 	}
 
-	/*
-	 * Keep fixed-link and real PHY configurations working, but allow an
+	/* Keep fixed-link and real PHY configurations working, but allow an
 	 * xPON-managed GDM2 to run without either. In the latter case the xPON
 	 * provider owns carrier, speed and duplex.
 	 */
-	err = airoha_gdm_phylink_connect(&dev->common,
+	err = airoha_gdm_phylink_connect(dev,
 					 dev->flags & AIROHA_PRIV_F_XPON_MANAGED);
 	if (err) {
 		netdev_err(netdev, "could not attach PHY: %d\n", err);
 		return err;
 	}
-	if (!dev->common.phylink_started &&
+	if (!dev->phylink_started &&
 	    (dev->flags & AIROHA_PRIV_F_XPON_MANAGED))
 		netdev_dbg(netdev,
 			   "no PHY or fixed-link, using xPON link state\n");
@@ -6734,11 +6406,7 @@ static int airoha_dev_open(struct net_device *netdev)
 		unsigned long flags;
 		bool link;
 
-		/*
-		 * Forcing carrier off here loses an O5 link that came up before
-		 * the interface was opened, and nothing re-announces it: adopt
-		 * the state the xPON provider already published.
-		 */
+		/* Adopt a link state already published by the xPON provider. */
 		spin_lock_irqsave(&dev->xpon_state_lock, flags);
 		link = dev->xpon_link.valid && dev->xpon_link.link;
 		spin_unlock_irqrestore(&dev->xpon_state_lock, flags);
@@ -6749,37 +6417,78 @@ static int airoha_dev_open(struct net_device *netdev)
 			netif_carrier_off(netdev);
 	}
 
-	netif_tx_start_all_queues(netdev);
-	err = airoha_set_vip_for_gdm_port(dev, true);
-	if (err)
-		return err;
+	if (econet) {
+		bool dsa = netdev_uses_dsa(netdev);
+		struct gdm_len_th rlt;
+		u16 oversize_len;
 
-	/* It seems GDM3 and GDM4 needs SPORT enabled to correctly work */
-	if (netdev_uses_dsa(netdev) || port->id > 2)
-		airoha_fe_set(eth, REG_GDM_INGRESS_CFG(port->id),
-			      GDM_STAG_EN_MASK);
-	else
-		airoha_fe_clear(eth, REG_GDM_INGRESS_CFG(port->id),
-				GDM_STAG_EN_MASK);
+		/* Keep the MTK special tag on a DSA conduit and disable it on a
+		 * direct PHY/WAN GDM. EN751221 additionally needs the legacy
+		 * CDM/GDM special-tag bits used by the vendor datapath.
+		 */
+		scoped_guard(spinlock, &dev->reg_lock) {
+			if (dev->fport == ETX_FPORT_GDM1) {
+				struct fwd_cfg fc;
+
+				fc = econet_rreg(&dev->econet_regs->fwd_cfg);
+				if (dsa) {
+					fc.word &= ~EN751221_GDM_UNTAG_EN;
+					fc.word |= EN751221_GDM_STAG_EN;
+				} else {
+					fc.word &= ~EN751221_GDM_STAG_EN;
+				}
+				econet_wreg(fc, &dev->econet_regs->fwd_cfg);
 
-	airoha_dev_set_mtu(netdev);
-	port->users++;
+				/* CDMA_CSG_CFG is at offset 0 from the GDM1 window. */
+				airoha_rmw(dev->regs, 0, EN751221_CDM_STAG_EN,
+					   dsa ? EN751221_CDM_STAG_EN : 0);
+			}
 
-	airoha_qdma_start(qdma);
+			econet_wreg((u32)dsa, &dev->econet_regs->stag_en);
+			rlt = econet_rreg(&dev->econet_regs->rx_len_threshold);
+			oversize_len = econet_gdm_oversize_len(dev, netdev->mtu);
+			set_gdm_len_th_runt_len(&rlt, 60);
+			set_gdm_len_th_oversize_len(&rlt, oversize_len);
+			econet_wreg(rlt, &dev->econet_regs->rx_len_threshold);
+		}
 
-	if (!airoha_is_lan_gdm_dev(dev) &&
-	    airoha_ppe_is_enabled(eth, 1))
-		pse_port = FE_PSE_PORT_PPE2;
-	airoha_set_gdm_port_fwd_cfg(eth, REG_GDM_FWD_CFG(port->id),
-				    pse_port);
+		airoha_qdma_start(qdma);
+		netif_tx_start_all_queues(netdev);
+	} else {
+		netif_tx_start_all_queues(netdev);
+		err = airoha_set_vip_for_gdm_port(dev, true);
+		if (err) {
+			netif_tx_disable(netdev);
+			airoha_gdm_phylink_disconnect(dev);
+			return err;
+		}
 
-	/* airoha_qdma_start() above already counted this device in. */
-	if ((netdev->features & NETIF_F_GRO_HW) && qdma->users <= 1)
-		airoha_dev_lro_enable(dev);
+		/* GDM3 and GDM4 need SPORT enabled to correctly work. */
+		if (netdev_uses_dsa(netdev) || port->id > 2)
+			airoha_fe_set(eth, REG_GDM_INGRESS_CFG(port->id),
+				      GDM_STAG_EN_MASK);
+		else
+			airoha_fe_clear(eth, REG_GDM_INGRESS_CFG(port->id),
+					GDM_STAG_EN_MASK);
+
+		airoha_dev_set_mtu(netdev);
+		port->users++;
+		airoha_qdma_start(qdma);
+
+		if (!airoha_is_lan_gdm_dev(dev) &&
+		    airoha_ppe_is_enabled(eth, 1))
+			pse_port = FE_PSE_PORT_PPE2;
+		airoha_set_gdm_port_fwd_cfg(eth, REG_GDM_FWD_CFG(port->id),
+					    pse_port);
 
-	airoha_update_netdev_features(dev);
+		/* airoha_qdma_start() above already counted this device in. */
+		if ((netdev->features & NETIF_F_GRO_HW) && qdma->users <= 1)
+			airoha_dev_lro_enable(dev);
 
-	err = airoha_eth_xpon_start(dev);
+		airoha_update_netdev_features(dev);
+	}
+
+	err = airoha_gdm_xpon_start(dev);
 	if (err) {
 		netdev_err(netdev, "failed to start xPON provider: %d\n", err);
 		airoha_dev_stop(netdev);
@@ -6793,35 +6502,54 @@ static int airoha_dev_stop(struct net_device *netdev)
 {
 	struct airoha_gdm_dev *dev = netdev_priv(netdev);
 	struct airoha_gdm_port *port = dev->port;
+	bool econet = airoha_is_econet(dev->eth);
 	struct airoha_qdma *qdma;
-	bool xpon_control;
-	int i;
 
-	airoha_eth_xpon_stop(dev);
-	netif_carrier_off(netdev);
+	airoha_gdm_xpon_stop(dev);
 	netif_tx_disable(netdev);
-	xpon_control = READ_ONCE(dev->xpon_control_started);
-	if (!xpon_control)
-		airoha_set_vip_for_gdm_port(dev, false);
-	for (i = 0; i < netdev->num_tx_queues; i++)
-		netdev_tx_reset_subqueue(netdev, i);
-
 	qdma = airoha_qdma_deref(dev);
-	if (--port->users)
-		airoha_ppe_set_mtu(dev);
-	else if (!xpon_control)
-		airoha_set_gdm_port_fwd_cfg(qdma->eth,
-					    REG_GDM_FWD_CFG(port->id),
-					    FE_PSE_PORT_DROP);
-	airoha_qdma_stop(qdma);
-	airoha_update_netdev_features(dev);
 
-	airoha_gdm_phylink_disconnect(&dev->common);
+	if (econet) {
+		scoped_guard(spinlock, &dev->reg_lock) {
+			econet_wreg(0U, &dev->econet_regs->stag_en);
+			if (dev->fport == ETX_FPORT_GDM1) {
+				struct fwd_cfg fc;
+
+				fc = econet_rreg(&dev->econet_regs->fwd_cfg);
+				fc.word &= ~EN751221_GDM_STAG_EN;
+				econet_wreg(fc, &dev->econet_regs->fwd_cfg);
+				airoha_rmw(dev->regs, 0, EN751221_CDM_STAG_EN, 0);
+			}
+		}
+
+		airoha_qdma_stop(qdma);
+	} else {
+		bool xpon_control;
+		int i;
+
+		netif_carrier_off(netdev);
+		xpon_control = READ_ONCE(dev->xpon_control_started);
+		if (!xpon_control)
+			airoha_set_vip_for_gdm_port(dev, false);
+		for (i = 0; i < netdev->num_tx_queues; i++)
+			netdev_tx_reset_subqueue(netdev, i);
+
+		if (--port->users)
+			airoha_ppe_set_mtu(dev);
+		else if (!xpon_control)
+			airoha_set_gdm_port_fwd_cfg(qdma->eth,
+						    REG_GDM_FWD_CFG(port->id),
+						    FE_PSE_PORT_DROP);
+		airoha_qdma_stop(qdma);
+		airoha_update_netdev_features(dev);
+	}
+
+	airoha_gdm_phylink_disconnect(dev);
 
 	return 0;
 }
 
-static int airoha_dev_set_macaddr(struct net_device *netdev, void *p)
+static int airoha_gdm_dev_set_macaddr(struct net_device *netdev, void *p)
 {
 	struct airoha_gdm_dev *dev = netdev_priv(netdev);
 	struct sockaddr *addr = p;
@@ -7015,6 +6743,43 @@ static int airoha_dev_init(struct net_device *netdev)
 {
 	struct airoha_gdm_dev *dev = netdev_priv(netdev);
 	struct airoha_gdm_port *port = dev->port;
+	bool econet = airoha_is_econet(dev->eth);
+	int err;
+
+	if (econet) {
+		enum etx_fport cpu_fport;
+
+		err = airoha_set_macaddr(dev, netdev->dev_addr);
+		if (err)
+			return err;
+
+		/* Route each GDM port to the CPU side of its matching QDMA. */
+		cpu_fport = dev->fport == ETX_FPORT_GDM2 ?
+			    ETX_FPORT_QDMA1_CPU : ETX_FPORT_QDMA0_CPU;
+		econet_set_gdm_port_fwd_cfg(dev, cpu_fport);
+
+		if (dev->fport == ETX_FPORT_GDM1) {
+			struct g1_cport_cfg cport_cfg;
+			struct gdm_vlan vlan;
+
+			/* Match macSetMACCR() in the EN7512 vendor driver. CDMA1
+			 * needs the 0x8100 insertion TPID and GDM1 CPORT padding
+			 * before LAN traffic is started.
+			 */
+			scoped_guard(spinlock, &dev->reg_lock) {
+				vlan = econet_rreg(&dev->econet_regs->vlan);
+				vlan.tpid = ETH_P_8021Q;
+				econet_wreg(vlan, &dev->econet_regs->vlan);
+
+				cport_cfg = econet_rreg(&dev->econet_regs->g1_cport_cfg);
+				set_gdm_g1_cport_cfg_pad(&cport_cfg, true);
+				econet_wreg(cport_cfg,
+					    &dev->econet_regs->g1_cport_cfg);
+			}
+		}
+
+		return 0;
+	}
 
 	switch (port->id) {
 	case AIROHA_GDM3_IDX:
@@ -7023,7 +6788,7 @@ static int airoha_dev_init(struct net_device *netdev)
 			break;
 		fallthrough;
 	case AIROHA_GDM2_IDX:
-		/* GDM2 is always used as wan */
+		/* GDM2 is always used as wan. */
 		dev->flags |= AIROHA_PRIV_F_WAN;
 		break;
 	default:
@@ -7031,50 +6796,58 @@ static int airoha_dev_init(struct net_device *netdev)
 	}
 
 	airoha_dev_set_qdma(dev);
-	airoha_set_macaddr(dev, netdev->dev_addr);
+	err = airoha_set_macaddr(dev, netdev->dev_addr);
+	if (err)
+		return err;
 
 	if (!airoha_is_lan_gdm_dev(dev) &&
-	    (port->id == AIROHA_GDM3_IDX || port->id == AIROHA_GDM4_IDX)) {
-		int err;
-
-		err = airoha_enable_gdm2_loopback(dev);
-		if (err)
-			return err;
-	}
+	    (port->id == AIROHA_GDM3_IDX || port->id == AIROHA_GDM4_IDX))
+		return airoha_enable_gdm2_loopback(dev);
 
 	return 0;
 }
 
-static void airoha_dev_get_stats64(struct net_device *netdev,
-				   struct rtnl_link_stats64 *storage)
+static void airoha_gdm_dev_get_stats64(struct net_device *netdev,
+				       struct rtnl_link_stats64 *storage)
 {
 	struct airoha_gdm_dev *dev = netdev_priv(netdev);
 	unsigned int start;
 
-	airoha_update_hw_stats(dev);
+	airoha_gdm_update_hw_stats(dev);
 	do {
 		start = u64_stats_fetch_begin(&dev->stats.syncp);
 		storage->rx_packets = dev->stats.rx_ok_pkts;
 		storage->tx_packets = dev->stats.tx_ok_pkts;
 		storage->rx_bytes = dev->stats.rx_ok_bytes;
 		storage->tx_bytes = dev->stats.tx_ok_bytes;
-		storage->multicast = dev->stats.rx_multicast;
 		storage->rx_errors = dev->stats.rx_errors;
 		storage->rx_dropped = dev->stats.rx_drops;
 		storage->tx_dropped = dev->stats.tx_drops;
-		storage->rx_crc_errors = dev->stats.rx_crc_error;
 		storage->rx_over_errors = dev->stats.rx_over_errors;
+		if (airoha_gdm_has_extended_stats(dev)) {
+			storage->multicast = dev->stats.rx_multicast;
+			storage->rx_crc_errors = dev->stats.rx_crc_error;
+		}
 	} while (u64_stats_fetch_retry(&dev->stats.syncp, start));
 }
 
-static int airoha_dev_change_mtu(struct net_device *netdev, int mtu)
+static int airoha_gdm_dev_change_mtu(struct net_device *netdev, int mtu)
 {
 	struct airoha_gdm_dev *dev = netdev_priv(netdev);
-	struct airoha_gdm_port *port = dev->port;
 
 	WRITE_ONCE(netdev->mtu, mtu);
-	if (port->users)
+
+	if (airoha_is_econet(dev->eth)) {
+		struct gdm_len_th rlt;
+
+		guard(spinlock)(&dev->reg_lock);
+		rlt = econet_rreg(&dev->econet_regs->rx_len_threshold);
+		set_gdm_len_th_oversize_len(&rlt,
+					    econet_gdm_oversize_len(dev, mtu));
+		econet_wreg(rlt, &dev->econet_regs->rx_len_threshold);
+	} else if (dev->port->users) {
 		airoha_dev_set_mtu(netdev);
+	}
 
 	return 0;
 }
@@ -7087,13 +6860,20 @@ static u16 airoha_dev_select_queue(struct net_device *netdev,
 	struct airoha_gdm_port *port = dev->port;
 	int queue, channel;
 
-	/* For dsa device select QoS channel according to the dsa user port
-	 * index, rely on port id otherwise. Select QoS queue based on the
-	 * skb priority.
+	/* EcoNet previously had no ndo_select_queue. Preserve the generic
+	 * networking-core queue selection now that all GDM netdevs share one
+	 * net_device_ops table.
+	 */
+	if (airoha_is_econet(dev->eth))
+		return netdev_pick_tx(netdev, skb, sb_dev);
+
+	/* For DSA devices select the QoS channel according to the DSA user port
+	 * index, rely on port id otherwise. Select QoS queue based on skb
+	 * priority.
 	 */
 	channel = netdev_uses_dsa(netdev) ? skb_get_queue_mapping(skb) : port->id;
 	channel = channel % AIROHA_NUM_QOS_CHANNELS;
-	queue = (skb->priority - 1) % AIROHA_NUM_QOS_QUEUES; /* QoS queue */
+	queue = (skb->priority - 1) % AIROHA_NUM_QOS_QUEUES;
 	queue = channel * AIROHA_NUM_QOS_QUEUES + queue;
 
 	return queue % dev->eth->soc->tx_ring;
@@ -7108,11 +6888,17 @@ int airoha_get_fe_port(struct airoha_gdm_dev *dev)
 }
 
 static netdev_features_t airoha_dev_fix_features(struct net_device *netdev,
-						 netdev_features_t features)
+					 netdev_features_t features)
 {
 	struct airoha_gdm_dev *dev = netdev_priv(netdev);
 	struct airoha_qdma *qdma;
 
+	/* EcoNet had no ndo_fix_features callback. Keep its feature set
+	 * unchanged; HW LRO/GRO programming below belongs to the modern FE.
+	 */
+	if (airoha_is_econet(dev->eth))
+		return features;
+
 	qdma = airoha_qdma_deref(dev);
 	if (qdma->users > 1)
 		features &= ~NETIF_F_GRO_HW;
@@ -7125,12 +6911,20 @@ static int airoha_dev_set_features(struct net_device *netdev,
 {
 	netdev_features_t diff = netdev->features ^ features;
 	struct airoha_gdm_dev *dev = netdev_priv(netdev);
-	struct airoha_qdma *qdma = airoha_qdma_deref(dev);
-	int qdma_id = qdma - &dev->eth->qdma[0];
+	struct airoha_qdma *qdma;
+	int qdma_id;
+
+	/* EcoNet had no ndo_set_features callback and does not use the modern
+	 * FE LRO registers.
+	 */
+	if (airoha_is_econet(dev->eth))
+		return 0;
 
 	if (!(diff & NETIF_F_GRO_HW))
 		return 0;
 
+	qdma = airoha_qdma_deref(dev);
+	qdma_id = qdma - &dev->eth->qdma[0];
 	if (features & NETIF_F_GRO_HW)
 		airoha_dev_lro_enable(dev);
 	else
@@ -7579,6 +7373,11 @@ static netdev_tx_t __airoha_dev_xmit(struct sk_buff *skb,
 static netdev_tx_t airoha_dev_xmit(struct sk_buff *skb,
 				   struct net_device *netdev)
 {
+	struct airoha_gdm_dev *dev = netdev_priv(netdev);
+
+	if (airoha_is_econet(dev->eth))
+		return econet_qdma_xmit(skb, netdev);
+
 	return __airoha_dev_xmit(skb, netdev, NULL);
 }
 
@@ -7618,16 +7417,18 @@ static void airoha_ethtool_get_mac_stats(struct net_device *netdev,
 	struct airoha_gdm_dev *dev = netdev_priv(netdev);
 	unsigned int start;
 
-	airoha_update_hw_stats(dev);
+	airoha_gdm_update_hw_stats(dev);
 	do {
 		start = u64_stats_fetch_begin(&dev->stats.syncp);
 		stats->FramesTransmittedOK = dev->stats.tx_ok_pkts;
 		stats->OctetsTransmittedOK = dev->stats.tx_ok_bytes;
-		stats->MulticastFramesXmittedOK = dev->stats.tx_multicast;
-		stats->BroadcastFramesXmittedOK = dev->stats.tx_broadcast;
 		stats->FramesReceivedOK = dev->stats.rx_ok_pkts;
 		stats->OctetsReceivedOK = dev->stats.rx_ok_bytes;
-		stats->BroadcastFramesReceivedOK = dev->stats.rx_broadcast;
+		if (airoha_gdm_has_extended_stats(dev)) {
+			stats->MulticastFramesXmittedOK = dev->stats.tx_multicast;
+			stats->BroadcastFramesXmittedOK = dev->stats.tx_broadcast;
+			stats->BroadcastFramesReceivedOK = dev->stats.rx_broadcast;
+		}
 	} while (u64_stats_fetch_retry(&dev->stats.syncp, start));
 }
 
@@ -7657,7 +7458,10 @@ airoha_ethtool_get_rmon_stats(struct net_device *netdev,
 		     ARRAY_SIZE(hw_stats->rx_len) + 1);
 
 	*ranges = airoha_ethtool_rmon_ranges;
-	airoha_update_hw_stats(dev);
+	airoha_gdm_update_hw_stats(dev);
+	if (!airoha_gdm_has_extended_stats(dev))
+		return;
+
 	do {
 		int i;
 
@@ -7680,8 +7484,8 @@ airoha_ethtool_get_link_ksettings(struct net_device *netdev,
 	struct airoha_xpon_link_state state;
 	unsigned long flags;
 
-	if (!(dev->flags & AIROHA_PRIV_F_XPON_MANAGED))
-		return phylink_ethtool_ksettings_get(dev->common.phylink, cmd);
+	if (!(READ_ONCE(dev->flags) & AIROHA_PRIV_F_XPON_MANAGED))
+		return phylink_ethtool_ksettings_get(dev->phylink, cmd);
 
 	spin_lock_irqsave(&dev->xpon_state_lock, flags);
 	state = dev->xpon_link;
@@ -7692,8 +7496,8 @@ airoha_ethtool_get_link_ksettings(struct net_device *netdev,
 	 * provider publishes its first state. PHY-less ports are reported
 	 * directly from the provider and do not depend on phylink.
 	 */
-	if (!state.valid && dev->common.phylink_started)
-		return phylink_ethtool_ksettings_get(dev->common.phylink, cmd);
+	if (!state.valid && dev->phylink_started)
+		return phylink_ethtool_ksettings_get(dev->phylink, cmd);
 
 	ethtool_link_ksettings_zero_link_mode(cmd, supported);
 	ethtool_link_ksettings_zero_link_mode(cmd, advertising);
@@ -7718,10 +7522,20 @@ airoha_ethtool_set_link_ksettings(struct net_device *netdev,
 {
 	struct airoha_gdm_dev *dev = netdev_priv(netdev);
 
-	if (dev->flags & AIROHA_PRIV_F_XPON_MANAGED)
+	if (READ_ONCE(dev->flags) & AIROHA_PRIV_F_XPON_MANAGED)
+		return -EOPNOTSUPP;
+
+	return phylink_ethtool_ksettings_set(dev->phylink, cmd);
+}
+
+static int airoha_ethtool_nway_reset(struct net_device *netdev)
+{
+	struct airoha_gdm_dev *dev = netdev_priv(netdev);
+
+	if (READ_ONCE(dev->flags) & AIROHA_PRIV_F_XPON_MANAGED)
 		return -EOPNOTSUPP;
 
-	return phylink_ethtool_ksettings_set(dev->common.phylink, cmd);
+	return phylink_ethtool_nway_reset(dev->phylink);
 }
 
 static int airoha_qdma_set_chan_tx_sched(struct net_device *netdev,
@@ -8576,17 +8390,23 @@ static int airoha_tc_setup_qdisc_htb(struct net_device *netdev,
 	return 0;
 }
 
-static int airoha_dev_tc_setup(struct net_device *dev,
-			       enum tc_setup_type type, void *type_data)
+static int airoha_gdm_dev_tc_setup(struct net_device *netdev,
+				   enum tc_setup_type type, void *type_data)
 {
+	struct airoha_gdm_dev *dev = netdev_priv(netdev);
+
+	if (airoha_is_econet(dev->eth))
+		return airoha_ppe_dev_setup_tc(dev->eth->ppe_dev, netdev, type,
+					       type_data);
+
 	switch (type) {
 	case TC_SETUP_QDISC_ETS:
-		return airoha_tc_setup_qdisc_ets(dev, type_data);
+		return airoha_tc_setup_qdisc_ets(netdev, type_data);
 	case TC_SETUP_QDISC_HTB:
-		return airoha_tc_setup_qdisc_htb(dev, type_data);
+		return airoha_tc_setup_qdisc_htb(netdev, type_data);
 	case TC_SETUP_BLOCK:
 	case TC_SETUP_FT:
-		return airoha_dev_setup_tc_block(dev, type_data);
+		return airoha_dev_setup_tc_block(netdev, type_data);
 	default:
 		return -EOPNOTSUPP;
 	}
@@ -8596,22 +8416,31 @@ static const struct net_device_ops airoha_netdev_ops = {
 	.ndo_init		= airoha_dev_init,
 	.ndo_open		= airoha_dev_open,
 	.ndo_stop		= airoha_dev_stop,
-	.ndo_change_mtu		= airoha_dev_change_mtu,
+	.ndo_change_mtu		= airoha_gdm_dev_change_mtu,
 	.ndo_select_queue	= airoha_dev_select_queue,
 	.ndo_fix_features	= airoha_dev_fix_features,
 	.ndo_set_features	= airoha_dev_set_features,
 	.ndo_start_xmit		= airoha_dev_xmit,
-	.ndo_get_stats64        = airoha_dev_get_stats64,
-	.ndo_set_mac_address	= airoha_dev_set_macaddr,
-	.ndo_setup_tc		= airoha_dev_tc_setup,
+	.ndo_get_stats64	= airoha_gdm_dev_get_stats64,
+	.ndo_set_mac_address	= airoha_gdm_dev_set_macaddr,
+	.ndo_setup_tc		= airoha_gdm_dev_tc_setup,
 };
 
+struct airoha_gdm_dev *airoha_gdm_from_netdev(struct net_device *netdev)
+{
+	if (!netdev || netdev->netdev_ops != &airoha_netdev_ops)
+		return NULL;
+
+	return netdev_priv(netdev);
+}
+
 static const struct ethtool_ops airoha_ethtool_ops = {
 	.get_drvinfo		= airoha_eth_get_drvinfo,
-	.get_eth_mac_stats      = airoha_ethtool_get_mac_stats,
+	.get_eth_mac_stats	= airoha_ethtool_get_mac_stats,
 	.get_rmon_stats		= airoha_ethtool_get_rmon_stats,
 	.get_link_ksettings	= airoha_ethtool_get_link_ksettings,
 	.set_link_ksettings	= airoha_ethtool_set_link_ksettings,
+	.nway_reset		= airoha_ethtool_nway_reset,
 	.get_link		= ethtool_op_get_link,
 };
 
@@ -8668,16 +8497,16 @@ bool airoha_is_valid_gdm_dev(struct airoha_eth *eth,
 }
 
 /* Nothing to do in MAC, everything is handled in PCS. */
-static void airoha_mac_config(void *priv, unsigned int mode,
+static void airoha_mac_config(struct airoha_gdm_dev *dev, unsigned int mode,
 			      const struct phylink_link_state *state)
 {
 }
 
-static void airoha_mac_link_up(void *priv, struct phy_device *phy,
+static void airoha_mac_link_up(struct airoha_gdm_dev *dev,
+			       struct phy_device *phy,
 			       unsigned int mode, phy_interface_t interface,
 			       int speed, int duplex, bool tx_pause, bool rx_pause)
 {
-	struct airoha_gdm_dev *dev = priv;
 	struct airoha_gdm_port *port = dev->port;
 	struct airoha_eth *eth = dev->eth;
 	u32 frag_size_tx, frag_size_rx;
@@ -8731,17 +8560,11 @@ static void airoha_mac_link_up(void *priv, struct phy_device *phy,
 }
 
 /* Nothing to do in MAC, everything is handled in PCS. */
-static void airoha_mac_link_down(void *priv, unsigned int mode,
+static void airoha_mac_link_down(struct airoha_gdm_dev *dev, unsigned int mode,
 				 phy_interface_t interface)
 {
 }
 
-static const struct airoha_gdm_mac_ops airoha_gdm_mac_ops = {
-	.mac_config = airoha_mac_config,
-	.mac_link_up = airoha_mac_link_up,
-	.mac_link_down = airoha_mac_link_down,
-};
-
 static int airoha_fill_available_pcs(struct phylink_config *config,
 				     struct phylink_pcs **available_pcs,
 				     unsigned int num_possible_pcs)
@@ -8768,7 +8591,19 @@ static int airoha_setup_phylink(struct net_device *netdev)
 		return err;
 	}
 
-	config = &dev->common.phylink_config;
+	config = &dev->phylink_config;
+
+	/* GDM2 in internal mode is reserved for the xPON datapath. */
+	if (port->id == AIROHA_GDM2_IDX &&
+	    phy_mode == PHY_INTERFACE_MODE_INTERNAL) {
+		dev->flags |= AIROHA_PRIV_F_XPON_MANAGED;
+		netdev->features |= NETIF_F_GRO_HW;
+
+		/* Keep RX GRO enabled but perform xPON TX segmentation in software. */
+		netdev->hw_features &= ~AIROHA_XPON_TX_OFFLOAD_FEATURES;
+		netdev->features &= ~AIROHA_XPON_TX_OFFLOAD_FEATURES;
+		netdev->vlan_features &= ~AIROHA_XPON_TX_OFFLOAD_FEATURES;
+	}
 
 	/*
 	 * An internal GDM link does not have a PHY or PCS to attach.  This is
@@ -8805,7 +8640,7 @@ static int airoha_setup_phylink(struct net_device *netdev)
 					   MAC_2500FD | MAC_5000FD | MAC_10000FD;
 
 		err = fwnode_phylink_pcs_parse(dev_fwnode(config->dev), NULL,
-					       &dev->common.phylink_config.num_available_pcs);
+					       &dev->phylink_config.num_available_pcs);
 		if (err)
 			return err;
 
@@ -8824,18 +8659,18 @@ static int airoha_setup_phylink(struct net_device *netdev)
 		if (phy_mode != PHY_INTERFACE_MODE_SGMII &&
 		    phy_mode != PHY_INTERFACE_MODE_1000BASEX) {
 			__set_bit(PHY_INTERFACE_MODE_2500BASEX,
-				  dev->common.phylink_config.supported_interfaces);
+				  dev->phylink_config.supported_interfaces);
 			__set_bit(PHY_INTERFACE_MODE_10GBASER,
-				  dev->common.phylink_config.supported_interfaces);
+				  dev->phylink_config.supported_interfaces);
 			__set_bit(PHY_INTERFACE_MODE_USXGMII,
-				  dev->common.phylink_config.supported_interfaces);
+				  dev->phylink_config.supported_interfaces);
 		}
 
 		phy_interface_copy(config->pcs_interfaces,
 				   config->supported_interfaces);
 	}
 
-	return airoha_gdm_phylink_create(&dev->common, np, phy_mode);
+	return airoha_gdm_phylink_create(dev, np, phy_mode);
 }
 
 static int airoha_alloc_gdm_device(struct airoha_eth *eth,
@@ -8869,20 +8704,17 @@ static int airoha_alloc_gdm_device(struct airoha_eth *eth,
 		return -ENOMEM;
 	}
 
+	netdev->netdev_ops = &airoha_netdev_ops;
+	netdev->ethtool_ops = &airoha_ethtool_ops;
 	if (econet) {
-		netdev->netdev_ops = &econet_netdev_ops;
-		netdev->ethtool_ops = &econet_ethtool_ops;
 		netdev->max_mtu = SKB_WITH_OVERHEAD(AIROHA_MAX_PACKET_SIZE) -
 				  ETH_HLEN - ETH_FCS_LEN;
-		netdev->hw_features = NETIF_F_IP_CSUM | NETIF_F_RXCSUM |
-				      NETIF_F_IPV6_CSUM;
+		netdev->hw_features = AIROHA_ETH_HW_FEATURES;
 		if (eth->ppe_dev && eth->ppe_dev->enabled)
 			netdev->hw_features |= NETIF_F_HW_TC;
 		netdev->features |= netdev->hw_features;
 		netdev->vlan_features = netdev->hw_features;
 	} else {
-		netdev->netdev_ops = &airoha_netdev_ops;
-		netdev->ethtool_ops = &airoha_ethtool_ops;
 		netdev->max_mtu = AIROHA_MAX_MTU;
 		netdev->hw_features = AIROHA_HW_FEATURES | NETIF_F_GRO_HW;
 		netdev->features |= AIROHA_HW_FEATURES;
@@ -8916,18 +8748,14 @@ static int airoha_alloc_gdm_device(struct airoha_eth *eth,
 
 	netdev->dev.of_node = of_node_get(np);
 	dev = netdev_priv(netdev);
-	airoha_gdm_common_init(&dev->common, eth, netdev,
-			       econet ? AIROHA_ETH_FAMILY_ECONET :
-					AIROHA_ETH_FAMILY_AIROHA,
-			       port->id,
-			       econet ? (port->id == AIROHA_GDM2_IDX ?
-					 DPORT_GDMA2 : DPORT_GDMA1) : 0,
-			       dev, econet ? NULL : &airoha_gdm_mac_ops);
-	dev->common.ppe = eth->ppe_dev;
-	u64_stats_init(&dev->stats.syncp);
 	dev->port = port;
 	dev->eth = eth;
 	dev->nbq = nbq;
+	dev->pse_port = econet ?
+		(port->id == AIROHA_GDM2_IDX ? DPORT_GDMA2 : DPORT_GDMA1) : 0;
+	dev->phylink_config.dev = &netdev->dev;
+	dev->phylink_config.type = PHYLINK_NETDEV;
+	u64_stats_init(&dev->stats.syncp);
 	mutex_init(&dev->xpon_lock);
 	spin_lock_init(&dev->xpon_state_lock);
 	spin_lock_init(&dev->xpon_service_lock);
@@ -8943,15 +8771,6 @@ static int airoha_alloc_gdm_device(struct airoha_eth *eth,
 			dev->flags |= AIROHA_PRIV_F_WAN;
 	}
 
-	if (of_property_read_bool(np, "airoha,xpon-managed")) {
-		dev->flags |= AIROHA_PRIV_F_XPON_MANAGED;
-		netdev->features |= NETIF_F_GRO_HW;
-
-		/* Keep the validated xPON RX GRO path but do TX segmentation in SW. */
-		netdev->hw_features &= ~AIROHA_XPON_TX_OFFLOAD_FEATURES;
-		netdev->features &= ~AIROHA_XPON_TX_OFFLOAD_FEATURES;
-		netdev->vlan_features &= ~AIROHA_XPON_TX_OFFLOAD_FEATURES;
-	}
 	port->devs[index] = dev;
 
 	return airoha_setup_phylink(netdev);
@@ -9162,7 +8981,7 @@ static void airoha_eth_cleanup(struct airoha_eth *eth)
 				netdev = netdev_from_priv(dev);
 				if (netdev->reg_state == NETREG_REGISTERED)
 					unregister_netdev(netdev);
-				airoha_gdm_phylink_destroy(&dev->common);
+				airoha_gdm_phylink_destroy(dev);
 				of_node_put(netdev->dev.of_node);
 				netdev->dev.of_node = NULL;
 			}
@@ -9518,6 +9337,8 @@ static const struct airoha_eth_xpon_ops airoha_xpon_ops = {
 
 const struct airoha_eth_soc_data econet_en751221_soc_data = {
 	.version = econet_en751221,
+	.qdma_gen = AIROHA_QDMA_GEN1,
+	.mac_addr_mode = AIROHA_MAC_ADDR_GDM_MASK,
 	.xpon_ops = &econet_xpon_ops,
 	.num_ppe = 1,
 	.irq_banks = 1,
@@ -9538,6 +9359,7 @@ const struct airoha_eth_soc_data econet_en751221_soc_data = {
 
 const struct airoha_eth_soc_data econet_en7528_soc_data = {
 	.version = econet_en7528,
+	.qdma_gen = AIROHA_QDMA_GEN1_5,
 	.num_ppe = 1,
 	.irq_banks = 4,
 	.max_gdm_ports = 2,
@@ -9557,6 +9379,7 @@ const struct airoha_eth_soc_data econet_en7528_soc_data = {
 
 const struct airoha_eth_soc_data airoha_en7581_soc_data = {
 	.version = airoha_en7581,
+	.qdma_gen = AIROHA_QDMA_GEN2_PLUS,
 	.xsi_rsts_names = en7581_xsi_rsts_names,
 	.num_xsi_rsts = ARRAY_SIZE(en7581_xsi_rsts_names),
 	.num_ppe = 2,
@@ -9578,6 +9401,7 @@ const struct airoha_eth_soc_data airoha_en7581_soc_data = {
 
 const struct airoha_eth_soc_data airoha_an7583_soc_data = {
 	.version = airoha_an7583,
+	.qdma_gen = AIROHA_QDMA_GEN2_PLUS,
 	.xsi_rsts_names = an7583_xsi_rsts_names,
 	.num_xsi_rsts = ARRAY_SIZE(an7583_xsi_rsts_names),
 	.num_ppe = 1,
@@ -9599,6 +9423,7 @@ const struct airoha_eth_soc_data airoha_an7583_soc_data = {
 
 const struct airoha_eth_soc_data airoha_en7523_soc_data = {
 	.version = airoha_en7523,
+	.qdma_gen = AIROHA_QDMA_GEN2_PLUS,
 	.xpon_ops = &airoha_xpon_ops,
 	.xsi_rsts_names = en7523_xsi_rsts_names,
 	.num_xsi_rsts = ARRAY_SIZE(en7523_xsi_rsts_names),
@@ -9809,27 +9634,16 @@ static void airoha_eth_remove(struct platform_device *pdev)
 }
 
 static const struct airoha_eth_xpon_ops *
-airoha_eth_get_xpon_ops(struct net_device *netdev,
-			struct airoha_gdm_common **gdm_out)
+airoha_eth_get_xpon_ops(struct net_device *netdev)
 {
-	struct airoha_gdm_common *gdm;
-	struct airoha_eth *eth;
-
-	if (!netdev)
-		return NULL;
-
-	gdm = airoha_gdm_common_from_netdev(netdev);
-	if (!gdm || gdm->id != 2 || !gdm->eth || !gdm->eth->soc)
-		return NULL;
+	struct airoha_gdm_dev *dev;
 
-	eth = gdm->eth;
-	if (!eth->soc->xpon_ops)
+	dev = airoha_gdm_from_netdev(netdev);
+	if (!dev || !dev->port || dev->port->id != AIROHA_GDM2_IDX ||
+	    !dev->eth || !dev->eth->soc)
 		return NULL;
 
-	if (gdm_out)
-		*gdm_out = gdm;
-
-	return eth->soc->xpon_ops;
+	return dev->eth->soc->xpon_ops;
 }
 
 struct net_device *airoha_eth_get_xpon_netdev(void)
@@ -9838,7 +9652,7 @@ struct net_device *airoha_eth_get_xpon_netdev(void)
 
 	rtnl_lock();
 	for_each_netdev(&init_net, netdev) {
-		if (!airoha_eth_get_xpon_ops(netdev, NULL))
+		if (!airoha_eth_get_xpon_ops(netdev))
 			continue;
 
 		dev_hold(netdev);
@@ -9854,7 +9668,7 @@ EXPORT_SYMBOL_GPL(airoha_eth_get_xpon_netdev);
 int airoha_eth_set_xpon_mode(struct net_device *netdev,
 			     enum airoha_xpon_mode mode)
 {
-	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev, NULL);
+	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev);
 
 	return ops && ops->set_mode ? ops->set_mode(netdev, mode) : -EOPNOTSUPP;
 }
@@ -9863,7 +9677,7 @@ EXPORT_SYMBOL_GPL(airoha_eth_set_xpon_mode);
 int airoha_eth_set_xpon_datapath(struct net_device *netdev,
 				 enum airoha_xpon_mode mode, bool enable)
 {
-	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev, NULL);
+	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev);
 
 	return ops && ops->set_datapath ?
 		ops->set_datapath(netdev, mode, enable) : -EOPNOTSUPP;
@@ -9873,7 +9687,7 @@ EXPORT_SYMBOL_GPL(airoha_eth_set_xpon_datapath);
 int airoha_eth_set_xpon_tcont_channel(struct net_device *netdev,
 				      unsigned int channel, bool enable)
 {
-	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev, NULL);
+	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev);
 
 	return ops && ops->set_tcont_channel ?
 		ops->set_tcont_channel(netdev, channel, enable) : -EOPNOTSUPP;
@@ -9885,7 +9699,7 @@ int airoha_eth_register_xpon(struct net_device *netdev,
 			     const struct airoha_xpon_link_ops *link_ops,
 			     void *priv)
 {
-	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev, NULL);
+	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev);
 
 	return ops && ops->register_link ?
 		ops->register_link(netdev, mode, link_ops, priv) : -EOPNOTSUPP;
@@ -9896,7 +9710,7 @@ void airoha_eth_unregister_xpon(struct net_device *netdev,
 				const struct airoha_xpon_link_ops *link_ops,
 				void *priv)
 {
-	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev, NULL);
+	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev);
 
 	if (ops && ops->unregister_link)
 		ops->unregister_link(netdev, link_ops, priv);
@@ -9906,7 +9720,7 @@ EXPORT_SYMBOL_GPL(airoha_eth_unregister_xpon);
 void airoha_eth_xpon_update_link(struct net_device *netdev,
 				 const struct airoha_xpon_link_state *state)
 {
-	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev, NULL);
+	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev);
 
 	if (ops && ops->update_link)
 		ops->update_link(netdev, state);
@@ -9915,7 +9729,7 @@ EXPORT_SYMBOL_GPL(airoha_eth_xpon_update_link);
 
 int airoha_eth_xpon_control_start(struct net_device *netdev)
 {
-	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev, NULL);
+	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev);
 
 	return ops && ops->control_start ? ops->control_start(netdev) : -EOPNOTSUPP;
 }
@@ -9923,7 +9737,7 @@ EXPORT_SYMBOL_GPL(airoha_eth_xpon_control_start);
 
 void airoha_eth_xpon_control_stop(struct net_device *netdev)
 {
-	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev, NULL);
+	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev);
 
 	if (ops && ops->control_stop)
 		ops->control_stop(netdev);
@@ -9932,7 +9746,7 @@ EXPORT_SYMBOL_GPL(airoha_eth_xpon_control_stop);
 
 void airoha_eth_xpon_dump_oam_rx_state(struct net_device *netdev)
 {
-	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev, NULL);
+	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev);
 
 	if (ops && ops->dump_oam_rx_state)
 		ops->dump_oam_rx_state(netdev);
@@ -9942,7 +9756,7 @@ EXPORT_SYMBOL_GPL(airoha_eth_xpon_dump_oam_rx_state);
 int airoha_eth_register_xpon_oam(struct net_device *netdev,
 				 struct airoha_xpon_oam_handler *handler)
 {
-	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev, NULL);
+	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev);
 
 	return ops && ops->register_oam ?
 		ops->register_oam(netdev, handler) : -EOPNOTSUPP;
@@ -9952,7 +9766,7 @@ EXPORT_SYMBOL_GPL(airoha_eth_register_xpon_oam);
 void airoha_eth_unregister_xpon_oam(struct net_device *netdev,
 				    struct airoha_xpon_oam_handler *handler)
 {
-	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev, NULL);
+	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev);
 
 	if (ops && ops->unregister_oam)
 		ops->unregister_oam(netdev, handler);
@@ -9962,7 +9776,7 @@ EXPORT_SYMBOL_GPL(airoha_eth_unregister_xpon_oam);
 int airoha_eth_xmit_xpon_oam(struct net_device *netdev, struct sk_buff *skb,
 			     u8 channel, u16 gem_port_id)
 {
-	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev, NULL);
+	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev);
 
 	return ops && ops->xmit_oam ?
 		ops->xmit_oam(netdev, skb, channel, gem_port_id) : -EOPNOTSUPP;
@@ -9972,7 +9786,7 @@ EXPORT_SYMBOL_GPL(airoha_eth_xmit_xpon_oam);
 int airoha_eth_xpon_add_service(struct net_device *netdev,
 				const struct airoha_xpon_service_cfg *cfg)
 {
-	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev, NULL);
+	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev);
 
 	return ops && ops->add_service ?
 		ops->add_service(netdev, cfg) : -EOPNOTSUPP;
@@ -9983,7 +9797,7 @@ int airoha_eth_xpon_get_tx_info(struct net_device *netdev, bool vlan_valid,
 				u16 vlan_id, bool pcp_valid, u8 pcp,
 				struct airoha_xpon_tx_info *info)
 {
-	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev, NULL);
+	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev);
 
 	return ops && ops->get_tx_info ?
 		ops->get_tx_info(netdev, vlan_valid, vlan_id, pcp_valid, pcp, info) :
@@ -9993,7 +9807,7 @@ int airoha_eth_xpon_get_tx_info(struct net_device *netdev, bool vlan_valid,
 bool airoha_eth_xpon_del_service(struct net_device *netdev, u32 cookie,
 				 u16 *gem_port_id)
 {
-	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev, NULL);
+	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev);
 
 	return ops && ops->del_service ?
 		ops->del_service(netdev, cookie, gem_port_id) : false;
@@ -10002,7 +9816,7 @@ EXPORT_SYMBOL_GPL(airoha_eth_xpon_del_service);
 
 bool airoha_eth_xpon_has_gem_service(struct net_device *netdev, u16 gem_port_id)
 {
-	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev, NULL);
+	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev);
 
 	return ops && ops->has_gem_service ?
 		ops->has_gem_service(netdev, gem_port_id) : false;
@@ -10011,7 +9825,7 @@ EXPORT_SYMBOL_GPL(airoha_eth_xpon_has_gem_service);
 
 void airoha_eth_xpon_flush_services(struct net_device *netdev)
 {
-	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev, NULL);
+	const struct airoha_eth_xpon_ops *ops = airoha_eth_get_xpon_ops(netdev);
 
 	if (ops && ops->flush_services)
 		ops->flush_services(netdev);
diff --git a/drivers/net/ethernet/airoha/airoha_eth.h b/drivers/net/ethernet/airoha/airoha_eth.h
index c8b6105aaed4..a861cde5b02c 100644
--- a/drivers/net/ethernet/airoha/airoha_eth.h
+++ b/drivers/net/ethernet/airoha/airoha_eth.h
@@ -31,6 +31,7 @@ struct ethtool_drvinfo;
 struct sk_buff;
 struct airoha_ppe_dev;
 struct airoha_eth;
+struct airoha_gdm_dev;
 struct gdm;
 
 #define AIROHA_MTK_INVALID_CHANNEL		7
@@ -92,64 +93,6 @@ void airoha_qdma_skb_get_mtk_meta(struct sk_buff *skb,
 				  enum airoha_mtk_tag_mode mode,
 				  struct airoha_qdma_skb_meta *meta);
 
-enum airoha_eth_family {
-	AIROHA_ETH_FAMILY_AIROHA,
-	AIROHA_ETH_FAMILY_ECONET,
-};
-
-/**
- * struct airoha_gdm_mac_ops - SoC-specific MAC callbacks
- * @mac_config: optional phylink MAC configuration callback
- * @mac_link_up: optional phylink link-up callback
- * @mac_link_down: optional phylink link-down callback
- *
- * GDM1 and GDM2 are functionally equivalent on the EcoNet and Airoha frame
- * engines.  The common object owns phylink and dispatches only the hardware
- * differences to the Ethernet backend.
- */
-struct airoha_gdm_mac_ops {
-	void (*mac_config)(void *priv, unsigned int mode,
-			   const struct phylink_link_state *state);
-	void (*mac_link_up)(void *priv, struct phy_device *phy,
-			    unsigned int mode, phy_interface_t interface,
-			    int speed, int duplex, bool tx_pause,
-			    bool rx_pause);
-	void (*mac_link_down)(void *priv, unsigned int mode,
-			      phy_interface_t interface);
-};
-
-#define AIROHA_GDM_COMMON_MAGIC	0x47444d43 /* "GDMC" */
-
-/**
- * struct airoha_gdm_common - common GDM netdev/phylink state
- * @magic: identifies a common GDM private object
- * @family: frame-engine family
- * @id: one-based GDM identifier
- * @pse_port: PPE/PSE destination port used by the backend
- * @eth: common frame-engine host
- * @netdev: associated Linux network device
- * @ppe: optional common PPE frontend
- * @priv: backend GDM object passed to @mac_ops
- * @mac_ops: optional SoC-specific MAC callbacks
- * @phylink: phylink instance
- * @phylink_config: phylink configuration owned by this GDM
- * @phylink_started: whether phylink_start() has been issued
- */
-struct airoha_gdm_common {
-	u32 magic;
-	enum airoha_eth_family family;
-	u8 id;
-	u8 pse_port;
-	struct airoha_eth *eth;
-	struct net_device *netdev;
-	struct airoha_ppe_dev *ppe;
-	void *priv;
-	const struct airoha_gdm_mac_ops *mac_ops;
-	struct phylink *phylink;
-	struct phylink_config phylink_config;
-	bool phylink_started;
-};
-
 u32 airoha_rr(void __iomem *base, u32 offset);
 void airoha_wr(void __iomem *base, u32 offset, u32 val);
 u32 airoha_rmw(void __iomem *base, u32 offset, u32 mask, u32 val);
@@ -163,26 +106,14 @@ int airoha_eth_init_mac_address(struct device *dev, struct device_node *np,
 void airoha_eth_get_drvinfo(struct net_device *netdev,
 			    struct ethtool_drvinfo *info);
 
-void airoha_gdm_common_init(struct airoha_gdm_common *gdm,
-			    struct airoha_eth *eth, struct net_device *netdev,
-			    enum airoha_eth_family family, u8 id,
-			    u8 pse_port, void *priv,
-			    const struct airoha_gdm_mac_ops *mac_ops);
-int airoha_gdm_phylink_create(struct airoha_gdm_common *gdm,
+struct airoha_gdm_dev *airoha_gdm_from_netdev(struct net_device *netdev);
+int airoha_gdm_phylink_create(struct airoha_gdm_dev *dev,
 			      struct device_node *np,
 			      phy_interface_t phy_mode);
-int airoha_gdm_phylink_connect(struct airoha_gdm_common *gdm,
+int airoha_gdm_phylink_connect(struct airoha_gdm_dev *dev,
 			       bool allow_no_phy);
-void airoha_gdm_phylink_disconnect(struct airoha_gdm_common *gdm);
-void airoha_gdm_phylink_destroy(struct airoha_gdm_common *gdm);
-
-static inline struct airoha_gdm_common *
-airoha_gdm_common_from_netdev(struct net_device *netdev)
-{
-	struct airoha_gdm_common *gdm = netdev_priv(netdev);
-
-	return gdm->magic == AIROHA_GDM_COMMON_MAGIC ? gdm : NULL;
-}
+void airoha_gdm_phylink_disconnect(struct airoha_gdm_dev *dev);
+void airoha_gdm_phylink_destroy(struct airoha_gdm_dev *dev);
 
 #define AIROHA_MAX_NUM_GDM_DEVS		3
 #define AIROHA_MAX_NUM_QDMA		2
@@ -247,10 +178,12 @@ airoha_gdm_common_from_netdev(struct net_device *netdev)
 #define AIROHA_RXQ_LRO_MAX_AGG_TIME		100
 #define AIROHA_RXQ_LRO_MAX_AGE_TIME		2000 /* 1ms */
 
+#define AIROHA_ETH_HW_FEATURES			\
+	(NETIF_F_IP_CSUM | NETIF_F_RXCSUM | NETIF_F_IPV6_CSUM)
+#define AIROHA_QDMA_TSO_HW_FEATURES		\
+	(NETIF_F_SG | NETIF_F_TSO | NETIF_F_TSO6)
 #define AIROHA_HW_FEATURES			\
-	(NETIF_F_IP_CSUM | NETIF_F_RXCSUM |	\
-	 NETIF_F_TSO6 | NETIF_F_IPV6_CSUM |	\
-	 NETIF_F_SG | NETIF_F_TSO | NETIF_F_HW_TC)
+	(AIROHA_ETH_HW_FEATURES | AIROHA_QDMA_TSO_HW_FEATURES | NETIF_F_HW_TC)
 
 #define PSE_RSV_PAGES			128
 #define PSE_QUEUE_RSV_PAGES		64
@@ -995,9 +928,6 @@ struct airoha_xpon_tx_info {
 };
 
 struct airoha_gdm_dev {
-	/* Must stay first: shared GDM/phylink and PPE port metadata. */
-	struct airoha_gdm_common common;
-
 	struct airoha_qdma __rcu *qdma;
 	struct airoha_gdm_port *port;
 	struct airoha_eth *eth;
@@ -1010,6 +940,7 @@ struct airoha_gdm_dev {
 	spinlock_t reg_lock;
 	bool g2_stats;
 	u8 fport;
+	u8 pse_port;
 
 	DECLARE_BITMAP(qos_sq_bmap, AIROHA_NUM_QOS_CHANNELS);
 	/* qos stats counters */
@@ -1047,6 +978,9 @@ struct airoha_gdm_dev {
 				 [AIROHA_XPON_SVC_MAX_COOKIES];
 	u8 xpon_service_ncookies[AIROHA_XPON_MAX_SERVICES];
 
+	struct phylink *phylink;
+	struct phylink_config phylink_config;
+	bool phylink_started;
 };
 
 struct airoha_gdm_port {
@@ -1174,8 +1108,22 @@ struct airoha_eth_xpon_ops {
 	void (*flush_services)(struct net_device *netdev);
 };
 
+enum airoha_mac_addr_mode {
+	AIROHA_MAC_ADDR_FE_RANGE,
+	AIROHA_MAC_ADDR_GDM_MASK,
+};
+
+enum airoha_qdma_generation {
+	AIROHA_QDMA_GEN1,
+	AIROHA_QDMA_GEN1_5,
+	AIROHA_QDMA_GEN2,
+	AIROHA_QDMA_GEN2_PLUS,
+};
+
 struct airoha_eth_soc_data {
 	enum airoha_ids version;
+	enum airoha_mac_addr_mode mac_addr_mode;
+	enum airoha_qdma_generation qdma_gen;
 	const struct airoha_eth_xpon_ops *xpon_ops;
 	const char * const *xsi_rsts_names;
 	int num_xsi_rsts;
@@ -1291,6 +1239,11 @@ static inline bool airoha_is_econet(struct airoha_eth *eth)
 	return airoha_is(eth, econet_en751221, econet_en7528);
 }
 
+static inline bool airoha_qdma_is_legacy(struct airoha_eth *eth)
+{
+	return eth->soc->qdma_gen <= AIROHA_QDMA_GEN1_5;
+}
+
 static inline bool airoha_qdma_is_lro_queue(struct airoha_queue *q)
 {
 	struct airoha_qdma *qdma = q->qdma;
diff --git a/drivers/net/ethernet/airoha/airoha_ppe.c b/drivers/net/ethernet/airoha/airoha_ppe.c
index ab8b2b6f2fd7..fd93d852c103 100644
--- a/drivers/net/ethernet/airoha/airoha_ppe.c
+++ b/drivers/net/ethernet/airoha/airoha_ppe.c
@@ -3371,20 +3371,20 @@ static void airoha_ppe_v1_engine_disarm(struct airoha_ppe *ppe)
 
 static struct airoha_ppe *airoha_ppe_from_netdev(struct net_device *netdev)
 {
-	struct airoha_gdm_common *gdm;
+	struct airoha_gdm_dev *gdm;
 
-	gdm = airoha_gdm_common_from_netdev(netdev);
-	if (!gdm || !gdm->ppe)
+	gdm = airoha_gdm_from_netdev(netdev);
+	if (!gdm || !gdm->eth || !gdm->eth->ppe_dev)
 		return NULL;
 
-	return gdm->ppe->priv;
+	return gdm->eth->ppe_dev->priv;
 }
 
 static int airoha_ppe_v1_flow_set_output(struct airoha_foe_entry *entry,
 				  struct net_device *odev, bool vlan_valid,
 				  u16 vlan_id, u8 vlan_prio, bool *xpon)
 {
-	struct airoha_gdm_common *gdm;
+	struct airoha_gdm_dev *gdm;
 	struct dsa_port *dp;
 	int dsa_port = -1;
 	int err;
@@ -3405,8 +3405,8 @@ static int airoha_ppe_v1_flow_set_output(struct airoha_foe_entry *entry,
 	if (!odev)
 		return -EOPNOTSUPP;
 
-	gdm = airoha_gdm_common_from_netdev(odev);
-	if (!gdm || gdm->family != AIROHA_ETH_FAMILY_ECONET)
+	gdm = airoha_gdm_from_netdev(odev);
+	if (!gdm || !gdm->eth || !airoha_is_econet(gdm->eth))
 		return -EOPNOTSUPP;
 
 	/*
-- 
2.55.0

